{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92e7a684-eb25-4cc3-b0e0-e3fe0435076f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import scprinter as scp\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "sc._settings.settings.n_jobs = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbb82e69-9e84-491e-8738-bb34fed79efd",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = '/ewsc/rzhang/GAGE_seq_AD/scATAC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3246927d-82dd-4b32-b87c-93628d40efa6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellId</th>\n",
       "      <th>Individual</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>msex</th>\n",
       "      <th>ADdiag3types</th>\n",
       "      <th>nFrags</th>\n",
       "      <th>FRIP</th>\n",
       "      <th>ReadsInPeaks</th>\n",
       "      <th>TSSEnrichment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D19-13182#CTTGCTGCATCCCAAA-1</td>\n",
       "      <td>ROSMAP-17378</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C11</td>\n",
       "      <td>0</td>\n",
       "      <td>earlyAD</td>\n",
       "      <td>99510</td>\n",
       "      <td>0.200988</td>\n",
       "      <td>39985</td>\n",
       "      <td>7.848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>D19-13182#TCGGGACTCACCACAA-1</td>\n",
       "      <td>ROSMAP-17378</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C11</td>\n",
       "      <td>0</td>\n",
       "      <td>earlyAD</td>\n",
       "      <td>99494</td>\n",
       "      <td>0.275523</td>\n",
       "      <td>54806</td>\n",
       "      <td>9.830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D19-13182#TGCTTTATCATGTTCT-1</td>\n",
       "      <td>ROSMAP-17378</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C15</td>\n",
       "      <td>0</td>\n",
       "      <td>earlyAD</td>\n",
       "      <td>96805</td>\n",
       "      <td>0.280257</td>\n",
       "      <td>54256</td>\n",
       "      <td>6.224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D19-13182#GCGAGTCAGTACCACT-1</td>\n",
       "      <td>ROSMAP-17378</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C11</td>\n",
       "      <td>0</td>\n",
       "      <td>earlyAD</td>\n",
       "      <td>95725</td>\n",
       "      <td>0.363587</td>\n",
       "      <td>69600</td>\n",
       "      <td>7.733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>D19-13182#GGAACCCTCTCCCTTT-1</td>\n",
       "      <td>ROSMAP-17378</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C12</td>\n",
       "      <td>0</td>\n",
       "      <td>earlyAD</td>\n",
       "      <td>94837</td>\n",
       "      <td>0.247250</td>\n",
       "      <td>46883</td>\n",
       "      <td>6.601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171114</th>\n",
       "      <td>D19-13184#TATTGCTAGTGATATG-1</td>\n",
       "      <td>ROSMAP-43882</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C8</td>\n",
       "      <td>1</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1012</td>\n",
       "      <td>0.322453</td>\n",
       "      <td>652</td>\n",
       "      <td>10.473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171115</th>\n",
       "      <td>D19-13184#TGAATCGAGAGACTCG-1</td>\n",
       "      <td>ROSMAP-43882</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Inh</td>\n",
       "      <td>C22</td>\n",
       "      <td>1</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1010</td>\n",
       "      <td>0.229663</td>\n",
       "      <td>463</td>\n",
       "      <td>10.608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171116</th>\n",
       "      <td>D19-13184#TATTGCTGTTCTTATC-1</td>\n",
       "      <td>ROSMAP-43882</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C15</td>\n",
       "      <td>1</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.173154</td>\n",
       "      <td>347</td>\n",
       "      <td>6.695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171117</th>\n",
       "      <td>D19-13184#CCAATGATCACATTCT-1</td>\n",
       "      <td>ROSMAP-43882</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1002</td>\n",
       "      <td>0.584830</td>\n",
       "      <td>1172</td>\n",
       "      <td>14.488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171118</th>\n",
       "      <td>D19-13184#CCCAGAGGTACCAAGG-1</td>\n",
       "      <td>ROSMAP-43882</td>\n",
       "      <td>Library11</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C6</td>\n",
       "      <td>1</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1000</td>\n",
       "      <td>0.280060</td>\n",
       "      <td>559</td>\n",
       "      <td>6.816</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171119 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              cellId    Individual      Batch Celltype  \\\n",
       "0       D19-13182#CTTGCTGCATCCCAAA-1  ROSMAP-17378  Library11      Ast   \n",
       "1       D19-13182#TCGGGACTCACCACAA-1  ROSMAP-17378  Library11      Ast   \n",
       "2       D19-13182#TGCTTTATCATGTTCT-1  ROSMAP-17378  Library11      Exc   \n",
       "3       D19-13182#GCGAGTCAGTACCACT-1  ROSMAP-17378  Library11      Ast   \n",
       "4       D19-13182#GGAACCCTCTCCCTTT-1  ROSMAP-17378  Library11      Ast   \n",
       "...                              ...           ...        ...      ...   \n",
       "171114  D19-13184#TATTGCTAGTGATATG-1  ROSMAP-43882  Library11      Oli   \n",
       "171115  D19-13184#TGAATCGAGAGACTCG-1  ROSMAP-43882  Library11      Inh   \n",
       "171116  D19-13184#TATTGCTGTTCTTATC-1  ROSMAP-43882  Library11      Exc   \n",
       "171117  D19-13184#CCAATGATCACATTCT-1  ROSMAP-43882  Library11      Oli   \n",
       "171118  D19-13184#CCCAGAGGTACCAAGG-1  ROSMAP-43882  Library11      Oli   \n",
       "\n",
       "       Cluster  msex ADdiag3types  nFrags      FRIP  ReadsInPeaks  \\\n",
       "0          C11     0      earlyAD   99510  0.200988         39985   \n",
       "1          C11     0      earlyAD   99494  0.275523         54806   \n",
       "2          C15     0      earlyAD   96805  0.280257         54256   \n",
       "3          C11     0      earlyAD   95725  0.363587         69600   \n",
       "4          C12     0      earlyAD   94837  0.247250         46883   \n",
       "...        ...   ...          ...     ...       ...           ...   \n",
       "171114      C8     1       lateAD    1012  0.322453           652   \n",
       "171115     C22     1       lateAD    1010  0.229663           463   \n",
       "171116     C15     1       lateAD    1003  0.173154           347   \n",
       "171117      C6     1       lateAD    1002  0.584830          1172   \n",
       "171118      C6     1       lateAD    1000  0.280060           559   \n",
       "\n",
       "        TSSEnrichment  \n",
       "0               7.848  \n",
       "1               9.830  \n",
       "2               6.224  \n",
       "3               7.733  \n",
       "4               6.601  \n",
       "...               ...  \n",
       "171114         10.473  \n",
       "171115         10.608  \n",
       "171116          6.695  \n",
       "171117         14.488  \n",
       "171118          6.816  \n",
       "\n",
       "[171119 rows x 11 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv('snATAC_meta_TSS_6.tsv', sep='\\t')\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "281fa6ea-64c3-4712-83f6-b49f3ac9eb50",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "lib2whitelist = {}\n",
    "for xx in meta['cellId']:\n",
    "    xx = xx.split(\"#\")\n",
    "    lib = xx[0]\n",
    "    bc = xx[1]\n",
    "    if lib not in lib2whitelist:\n",
    "        lib2whitelist[lib] = []\n",
    "    lib2whitelist[lib].append(bc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "950a6c68-f942-4354-9d7c-bde417197cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "libs = ['D19-8609', \n",
    "'D19-12698', \n",
    "'D19-12529', \n",
    "'D19-13171', \n",
    "'D19-12530', \n",
    "'D19-13184', \n",
    "'D19-8608', \n",
    "'D19-12525', \n",
    "'D19-12701', \n",
    "'D19-12513', \n",
    "'D19-13167', \n",
    "'D19-125460', \n",
    "'D19-8610', \n",
    "'D19-13154', \n",
    "'D19-125459', \n",
    "'D19-13177', \n",
    "'D19-12522', \n",
    "'D19-13175', \n",
    "'D19-125454', \n",
    "'D19-12527']\n",
    "frags = [f'./fragments/{lib}_fragments.bed.gz' for lib in libs]\n",
    "whitelists = [lib2whitelist[lib] for lib in libs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de9c9322-08c5-456c-a0aa-bbd2f4563992",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = '/ewsc/zhangruo/GAGE_seq_AD/scATAC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60807eb2-36de-440d-97f0-913f7e4004b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You are now using the beta auto_detect_shift function, this overwrites the plus_shift and minus_shift you provided\n",
      "If you believe the auto_detect_shift is wrong, please set auto_detect_shift=False\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "89aa8649f8304cea82d4f9de4cb15e05",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Importing fragments:   0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0017856514509375, shift detected\n",
      "Minimum MSE is 0.0017965175130125002, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-8609_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 9.554078218750006e-05, shift detected\n",
      "Minimum MSE is 0.00010659780676250002, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12698_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0001597581021875001, shift detected\n",
      "Minimum MSE is 0.00011379893176250004, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12529_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0012031947196875, shift detected\n",
      "Minimum MSE is 0.0011681575092625002, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13171_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0023647538871875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0023374097092625002\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12530_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0033616353634374997\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0034606930580124997\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13184_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0009654603296875002, shift detected\n",
      "Minimum MSE is 0.0009313972955125, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-8608_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0028854148146875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0028133065542625\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12525_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0026915504771875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0026229319080124996\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12701_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022665132709375002\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022837757455125\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12513_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0023519295921875004\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022663179280125\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13167_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0013248694271875, shift detected\n",
      "Minimum MSE is 0.0012755116667625, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-125460_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0018268772471875, shift detected\n",
      "Minimum MSE is 0.0017820704230125002, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-8610_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022721385509375\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022585636692625003\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13154_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0011783733246875001, shift detected\n",
      "Minimum MSE is 0.0012738549767625, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-125459_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0024375297571875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0024122853592624996\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13177_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0026306052121875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0026490226192625\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12522_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0021883396746875\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0022282820542624997\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-13175_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Minimum MSE is 0.0014854465171875004, shift detected\n",
      "Minimum MSE is 0.0014500452055124999, shift detected\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-125454_fragments.bed.gz\n",
      "Detecting the shift in the paired end fragments file\n",
      "If you think the above message is wrong, please check the input file format\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0034150592021874996\n",
      "Warning: The shift detection may not be accurate. The minimum MSE is 0.0033433019930125\n",
      "detected plus_shift and minus_shift are 4 -5 for ./fragments/D19-12527_fragments.bed.gz\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "291013a474a4419a9de90c2779a90ed3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start transferring insertions\n",
      "takes 1807.1912803649902\n"
     ]
    }
   ],
   "source": [
    "import scprinter as scp\n",
    "import time\n",
    "start = time.time()\n",
    "\n",
    "if os.path.exists(f'{work_dir}/scprinter_tss6.h5ad'):\n",
    "    printer = scp.load_printer(f'{work_dir}/scprinter_tss6.h5ad', scp.genome.hg38)\n",
    "else:\n",
    "    printer = scp.pp.import_fragments(\n",
    "                        path_to_frags=frags,\n",
    "                        barcodes=whitelists,\n",
    "                        savename=f'{work_dir}/scprinter_tss6.h5ad',\n",
    "                        genome=scp.genome.hg38,\n",
    "                        min_num_fragments=0, min_tsse=0,\n",
    "                        sorted_by_barcode=False, \n",
    "                        low_memory=False,\n",
    "                        sample_names=libs\n",
    "                        )\n",
    "print (\"takes\", time.time()-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2d192b8b-740e-45b4-93f6-5ce727b54e07",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import chain\n",
    "all_whitelist = []\n",
    "for lib, wl in zip(libs, whitelists):\n",
    "    for x in wl:\n",
    "        all_whitelist.append(f'{lib}_{x}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "71b2ec08-f6b1-429f-805d-f5116a12d24a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellId</th>\n",
       "      <th>Individual</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>msex</th>\n",
       "      <th>ADdiag3types</th>\n",
       "      <th>nFrags</th>\n",
       "      <th>FRIP</th>\n",
       "      <th>ReadsInPeaks</th>\n",
       "      <th>TSSEnrichment</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>D19-8609_ATTGTCTTCTGGCTAA-1</th>\n",
       "      <td>D19-8609#ATTGTCTTCTGGCTAA-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>99891</td>\n",
       "      <td>0.325869</td>\n",
       "      <td>64964</td>\n",
       "      <td>6.236</td>\n",
       "      <td>D19-8609_ATTGTCTTCTGGCTAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_TATGTTCTCTTCAGAG-1</th>\n",
       "      <td>D19-8609#TATGTTCTCTTCAGAG-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>98917</td>\n",
       "      <td>0.306965</td>\n",
       "      <td>60587</td>\n",
       "      <td>6.752</td>\n",
       "      <td>D19-8609_TATGTTCTCTTCAGAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_CATTCATTCGGAAAGC-1</th>\n",
       "      <td>D19-8609#CATTCATTCGGAAAGC-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>95683</td>\n",
       "      <td>0.312696</td>\n",
       "      <td>59730</td>\n",
       "      <td>6.747</td>\n",
       "      <td>D19-8609_CATTCATTCGGAAAGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_TCAGGGCAGCACGTAG-1</th>\n",
       "      <td>D19-8609#TCAGGGCAGCACGTAG-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Inh</td>\n",
       "      <td>C20</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>94457</td>\n",
       "      <td>0.288590</td>\n",
       "      <td>54443</td>\n",
       "      <td>7.978</td>\n",
       "      <td>D19-8609_TCAGGGCAGCACGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_GTGTCCTAGCCTTTGA-1</th>\n",
       "      <td>D19-8609#GTGTCCTAGCCTTTGA-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C10</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>93968</td>\n",
       "      <td>0.213309</td>\n",
       "      <td>39983</td>\n",
       "      <td>6.631</td>\n",
       "      <td>D19-8609_GTGTCCTAGCCTTTGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_CGGACCACAGTCAGCC-1</th>\n",
       "      <td>D19-12527#CGGACCACAGTCAGCC-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1080</td>\n",
       "      <td>0.421296</td>\n",
       "      <td>910</td>\n",
       "      <td>11.279</td>\n",
       "      <td>D19-12527_CGGACCACAGTCAGCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_ACTTTCAGTCTAACCA-1</th>\n",
       "      <td>D19-12527#ACTTTCAGTCTAACCA-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1075</td>\n",
       "      <td>0.141860</td>\n",
       "      <td>305</td>\n",
       "      <td>6.178</td>\n",
       "      <td>D19-12527_ACTTTCAGTCTAACCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_GATCATGGTTTCCGGG-1</th>\n",
       "      <td>D19-12527#GATCATGGTTTCCGGG-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1062</td>\n",
       "      <td>0.322976</td>\n",
       "      <td>686</td>\n",
       "      <td>9.424</td>\n",
       "      <td>D19-12527_GATCATGGTTTCCGGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_ATGTCTTAGGGTTCCC-1</th>\n",
       "      <td>D19-12527#ATGTCTTAGGGTTCCC-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1033</td>\n",
       "      <td>0.296709</td>\n",
       "      <td>613</td>\n",
       "      <td>8.086</td>\n",
       "      <td>D19-12527_ATGTCTTAGGGTTCCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_TGGGTTACAATCACTG-1</th>\n",
       "      <td>D19-12527#TGGGTTACAATCACTG-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C10</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1026</td>\n",
       "      <td>0.214425</td>\n",
       "      <td>440</td>\n",
       "      <td>8.628</td>\n",
       "      <td>D19-12527_TGGGTTACAATCACTG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25949 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    cellId    Individual  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    D19-8609#ATTGTCTTCTGGCTAA-1  ROSMAP-33422   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    D19-8609#TATGTTCTCTTCAGAG-1  ROSMAP-33422   \n",
       "D19-8609_CATTCATTCGGAAAGC-1    D19-8609#CATTCATTCGGAAAGC-1  ROSMAP-33422   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    D19-8609#TCAGGGCAGCACGTAG-1  ROSMAP-33422   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    D19-8609#GTGTCCTAGCCTTTGA-1  ROSMAP-33422   \n",
       "...                                                    ...           ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1  D19-12527#CGGACCACAGTCAGCC-1  ROSMAP-42915   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  D19-12527#ACTTTCAGTCTAACCA-1  ROSMAP-42915   \n",
       "D19-12527_GATCATGGTTTCCGGG-1  D19-12527#GATCATGGTTTCCGGG-1  ROSMAP-42915   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  D19-12527#ATGTCTTAGGGTTCCC-1  ROSMAP-42915   \n",
       "D19-12527_TGGGTTACAATCACTG-1  D19-12527#TGGGTTACAATCACTG-1  ROSMAP-42915   \n",
       "\n",
       "                                 Batch Celltype Cluster  msex ADdiag3types  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_CATTCATTCGGAAAGC-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1   Library2      Inh     C20     1        nonAD   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1   Library2      Ast     C10     1        nonAD   \n",
       "...                                ...      ...     ...   ...          ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_GATCATGGTTTCCGGG-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_TGGGTTACAATCACTG-1  Library7      Ast     C10     0       lateAD   \n",
       "\n",
       "                              nFrags      FRIP  ReadsInPeaks  TSSEnrichment  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    99891  0.325869         64964          6.236   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    98917  0.306965         60587          6.752   \n",
       "D19-8609_CATTCATTCGGAAAGC-1    95683  0.312696         59730          6.747   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    94457  0.288590         54443          7.978   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    93968  0.213309         39983          6.631   \n",
       "...                              ...       ...           ...            ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1    1080  0.421296           910         11.279   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1    1075  0.141860           305          6.178   \n",
       "D19-12527_GATCATGGTTTCCGGG-1    1062  0.322976           686          9.424   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1    1033  0.296709           613          8.086   \n",
       "D19-12527_TGGGTTACAATCACTG-1    1026  0.214425           440          8.628   \n",
       "\n",
       "                                                     index  \n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    D19-8609_ATTGTCTTCTGGCTAA-1  \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    D19-8609_TATGTTCTCTTCAGAG-1  \n",
       "D19-8609_CATTCATTCGGAAAGC-1    D19-8609_CATTCATTCGGAAAGC-1  \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    D19-8609_TCAGGGCAGCACGTAG-1  \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    D19-8609_GTGTCCTAGCCTTTGA-1  \n",
       "...                                                    ...  \n",
       "D19-12527_CGGACCACAGTCAGCC-1  D19-12527_CGGACCACAGTCAGCC-1  \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  D19-12527_ACTTTCAGTCTAACCA-1  \n",
       "D19-12527_GATCATGGTTTCCGGG-1  D19-12527_GATCATGGTTTCCGGG-1  \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  D19-12527_ATGTCTTAGGGTTCCC-1  \n",
       "D19-12527_TGGGTTACAATCACTG-1  D19-12527_TGGGTTACAATCACTG-1  \n",
       "\n",
       "[25949 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.index= [xx.split(\"#\")[0] + \"_\" + xx.split(\"#\")[1] for xx in meta['cellId']]\n",
    "meta['index'] = meta.index\n",
    "meta = meta.loc[all_whitelist].copy()\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c69ea56a-090c-4b0d-8760-1dfb2acc131e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_grouping, group_names = scp.utils.df2cell_grouping(printer, meta[['index', 'Celltype']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3c8dc663-e3f3-4358-99d7-987e1d264d90",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Mic_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Mic -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:03:18: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Mic_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Mic -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Mic\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Mic_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:18: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:18: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:19:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:20:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:21:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:22:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:23:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:24:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:25:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:26:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:27:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:28:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:30:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:31:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:32:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:33:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:34:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #1 mean fragment size is determined as 169.2 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #1 fragment size = 169.2 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #1  total fragments in treatment: 15644227 \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:44: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:52: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:52: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Mic_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:52: #3   Write bedGraph files for control lambda (after scaling if necessary)... Mic_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:52: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:03:52: #3 Call peaks for each chromosome... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Vas_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Vas -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:04:11: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Vas_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Vas -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Vas\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Vas_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:11: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:11: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:12:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #1 mean fragment size is determined as 167.5 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #1 fragment size = 167.5 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #1  total fragments in treatment: 1263231 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:13: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:14: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:14: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Vas_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:14: #3   Write bedGraph files for control lambda (after scaling if necessary)... Vas_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:14: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:14: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:19: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Vas_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:19: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Vas_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:19: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Vas_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:19: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:32: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Mic_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:33: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Mic_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:33: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Mic_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:33: Done! \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Opc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Opc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:04:35: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Opc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Opc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Opc\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Opc_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:35: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:35: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:36:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:37:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:38:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:39:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:40:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:41:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:42:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:44:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:45:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:46:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:47:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:48:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:49:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:50:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:51:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:52:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:04:53:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #1 mean fragment size is determined as 181.6 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #1 fragment size = 181.6 \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #1  total fragments in treatment: 17342457 \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:04: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:13: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:13: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Opc_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:13: #3   Write bedGraph files for control lambda (after scaling if necessary)... Opc_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:13: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:13: #3 Call peaks for each chromosome... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Ast_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Ast -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:05:57: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Ast_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Ast -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Ast\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Ast_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:57: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:57: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:58:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:59:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:05:59: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Opc_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:00: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Opc_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:00: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Opc_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:00:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:00: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:01:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:02:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:03:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:04:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:05:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:06:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:08:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:09:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:10:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:11:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:12:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:13:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:14:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:15:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:16:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:17:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:18:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:19:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:20:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:21:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:22:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:23:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:24:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:25:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:27:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:28:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:29:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:30:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:31:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:32:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:33:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:34:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:35:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:06:36:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #1 mean fragment size is determined as 171.5 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #1 fragment size = 171.5 \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #1  total fragments in treatment: 37760723 \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:01: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:19: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:19: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Ast_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:19: #3   Write bedGraph files for control lambda (after scaling if necessary)... Ast_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:19: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:07:19: #3 Call peaks for each chromosome... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Inh_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Inh -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:08:22: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Inh_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Inh -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Inh\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Inh_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:22: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:22: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:23:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:24:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:25:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:26:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:27:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:28:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:29: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Ast_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:29:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:29: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Ast_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:30: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Ast_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:30: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:30:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:31:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:32:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:33:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:34:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:35:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:36:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:38:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:39:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:40:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:41:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:42:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:43:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:44:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:45:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:46:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:47:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:48:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:49:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:50:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:51:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:52:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:53:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:54:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:55:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:56:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:58:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:08:59:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:00:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:01:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:02:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:03:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:04:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:05:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:06:  42000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:07:  43000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:08:  44000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:09:  45000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:10:  46000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:11:  47000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:12:  48000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:13:  49000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:14:  50000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:15:  51000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:17:  52000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:18:  53000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:19:  54000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:20:  55000000 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Exc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Exc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:09:40: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Exc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Exc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Exc\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Exc_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:40: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:40: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:41:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:42:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:43:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:44:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:45:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:46:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:47:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:48:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:49:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:50:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:51:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:53:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:54:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:55:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:56:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #1 mean fragment size is determined as 192.3 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #1 fragment size = 192.3 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #1  total fragments in treatment: 55976129 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:57: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:58:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:09:59:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:00:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:01:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:02:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:03:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:04:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:05:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:06:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:07:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:08:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:09:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:10:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:11:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:12:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:13:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:14:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:15:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:16:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:17:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:18:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:19:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:20:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:21:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:22:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23:  42000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Inh_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23: #3   Write bedGraph files for control lambda (after scaling if necessary)... Inh_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:23: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:24:  43000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:25:  44000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:26:  45000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:27:  46000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:29:  47000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:30:  48000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:31:  49000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:32:  50000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:33:  51000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:34:  52000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:35:  53000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:36:  54000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:37:  55000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:38:  56000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:39:  57000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:40:  58000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:41:  59000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:42:  60000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:43:  61000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:44:  62000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:45:  63000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:46:  64000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:47:  65000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:48:  66000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:10:49:  67000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #1 mean fragment size is determined as 195.0 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #1 fragment size = 195.0 \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #1  total fragments in treatment: 67083069 \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:11:33: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:03: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:03: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Exc_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:03: #3   Write bedGraph files for control lambda (after scaling if necessary)... Exc_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:03: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:03: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:05: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Inh_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:06: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Inh_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:06: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Inh_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:12:06: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 16:13:51: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Exc_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:13:52: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Exc_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:13:52: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Exc_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:13:52: Done! \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Oli_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Oli -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 16:19:00: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Oli_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n Oli -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = Oli\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/Oli_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:00: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:00: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:01:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:02:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:03:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:04:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:05:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:06:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:07:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:08:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:09:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:10:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:11:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:12:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:13:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:14:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:15:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:16:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:17:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:18:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:19:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:20:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:21:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:22:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:23:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:24:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:25:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:26:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:27:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:28:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:29:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:30:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:31:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:32:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:33:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:34:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:35:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:36:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:37:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:37:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:38:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:39:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:40:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:41:  42000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:42:  43000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:43:  44000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:44:  45000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:45:  46000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:46:  47000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:47:  48000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:48:  49000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:49:  50000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:50:  51000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:51:  52000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:52:  53000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:53:  54000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:54:  55000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:55:  56000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:56:  57000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:57:  58000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:58:  59000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:19:59:  60000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:00:  61000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:01:  62000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:02:  63000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:03:  64000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:04:  65000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:05:  66000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:06:  67000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:06:  68000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:07:  69000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:08:  70000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:09:  71000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:10:  72000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:11:  73000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:12:  74000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:13:  75000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:14:  76000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:15:  77000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:16:  78000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:17:  79000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:18:  80000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:19:  81000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:20:  82000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:21:  83000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:22:  84000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:23:  85000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:24:  86000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:25:  87000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:26:  88000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:27:  89000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:28:  90000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:29:  91000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:30:  92000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:31:  93000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:32:  94000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:33:  95000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:34:  96000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:34:  97000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:35:  98000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:36:  99000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:37:  100000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:38:  101000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:39:  102000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:40:  103000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:41:  104000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:42:  105000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:43:  106000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:44:  107000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:45:  108000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:46:  109000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:47:  110000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:48:  111000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:49:  112000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:50:  113000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:51:  114000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:52:  115000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:53:  116000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:54:  117000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:55:  118000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:56:  119000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:56:  120000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:57:  121000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:58:  122000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:20:59:  123000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:00:  124000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:01:  125000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:02:  126000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:03:  127000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:04:  128000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:05:  129000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:06:  130000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:07:  131000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:08:  132000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:09:  133000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:10:  134000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:11:  135000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:12:  136000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:13:  137000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:14:  138000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:15:  139000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:16:  140000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:21:17:  141000000 \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #1 mean fragment size is determined as 173.9 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #1 fragment size = 173.9 \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #1  total fragments in treatment: 141075095 \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 16:22:49: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 16:23:38: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 16:23:38: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... Oli_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:23:38: #3   Write bedGraph files for control lambda (after scaling if necessary)... Oli_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 16:23:38: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 16:23:38: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 16:25:47: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Oli_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 16:25:48: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Oli_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 16:25:48: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/Oli_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 16:25:48: Done! \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1a198d5035a74aee8837387f4f7cd40f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/7 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 1205342 peaks unresolved 311571 peaks selected\n",
      "round: 2 321417 peaks unresolved 65448 peaks selected\n",
      "round: 3 104062 peaks unresolved 19759 peaks selected\n",
      "round: 4 34704 peaks unresolved 6548 peaks selected\n",
      "round: 5 11159 peaks unresolved 2078 peaks selected\n",
      "round: 6 3664 peaks unresolved 682 peaks selected\n",
      "round: 7 1064 peaks unresolved 220 peaks selected\n",
      "round: 8 301 peaks unresolved 69 peaks selected\n",
      "round: 9 55 peaks unresolved 13 peaks selected\n",
      "round: 10 16 peaks unresolved 6 peaks selected\n",
      "round: 11 2 peaks unresolved 1 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n"
     ]
    }
   ],
   "source": [
    "scp.pp.call_peaks(\n",
    "    printer=printer,\n",
    "    frag_file=frags,\n",
    "    cell_grouping=cell_grouping,\n",
    "    group_names=list(group_names),\n",
    "    iterative_peak_merging=True,\n",
    "    merge_across_groups=True,\n",
    "    preset='chromvar',\n",
    "    sample_names=libs,\n",
    "    overwrite=False,\n",
    "    fdr_threshold=0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1bcc1caa-4ae0-4eb6-a3b7-ad71b9ee5a05",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3158716219.py:1: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling']['merged']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>final_score</th>\n",
       "      <th>cell</th>\n",
       "      <th>rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>817183</td>\n",
       "      <td>817483</td>\n",
       "      <td>Mic_peak_2</td>\n",
       "      <td>49.0784</td>\n",
       "      <td>49.0784</td>\n",
       "      <td>Mic</td>\n",
       "      <td>71146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818121</td>\n",
       "      <td>818421</td>\n",
       "      <td>Opc_peak_3</td>\n",
       "      <td>5.43633</td>\n",
       "      <td>5.43633</td>\n",
       "      <td>Opc</td>\n",
       "      <td>365799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>827386</td>\n",
       "      <td>827686</td>\n",
       "      <td>Oli_peak_4</td>\n",
       "      <td>331.329</td>\n",
       "      <td>331.329</td>\n",
       "      <td>Oli</td>\n",
       "      <td>15933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>850353</td>\n",
       "      <td>850653</td>\n",
       "      <td>Exc_peak_4</td>\n",
       "      <td>4.58036</td>\n",
       "      <td>4.58036</td>\n",
       "      <td>Exc</td>\n",
       "      <td>262170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869792</td>\n",
       "      <td>870092</td>\n",
       "      <td>Oli_peak_5</td>\n",
       "      <td>66.1156</td>\n",
       "      <td>66.1156</td>\n",
       "      <td>Oli</td>\n",
       "      <td>55696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406390</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155820136</td>\n",
       "      <td>155820436</td>\n",
       "      <td>Oli_peak_112984</td>\n",
       "      <td>11.1663</td>\n",
       "      <td>11.1663</td>\n",
       "      <td>Oli</td>\n",
       "      <td>176104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406391</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155845361</td>\n",
       "      <td>155845661</td>\n",
       "      <td>Ast_peak_112992</td>\n",
       "      <td>5.6359</td>\n",
       "      <td>5.6359</td>\n",
       "      <td>Ast</td>\n",
       "      <td>241527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406392</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155874756</td>\n",
       "      <td>155875056</td>\n",
       "      <td>Exc_peak_156151</td>\n",
       "      <td>10.639</td>\n",
       "      <td>10.639</td>\n",
       "      <td>Exc</td>\n",
       "      <td>179803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406393</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881107</td>\n",
       "      <td>155881407</td>\n",
       "      <td>Oli_peak_112985</td>\n",
       "      <td>413.806</td>\n",
       "      <td>413.806</td>\n",
       "      <td>Oli</td>\n",
       "      <td>13398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406394</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155956590</td>\n",
       "      <td>155956890</td>\n",
       "      <td>Ast_peak_112994</td>\n",
       "      <td>5.6359</td>\n",
       "      <td>5.6359</td>\n",
       "      <td>Ast</td>\n",
       "      <td>241287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>406395 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end             name    score final_score cell  \\\n",
       "0       chr1     817183     817483       Mic_peak_2  49.0784     49.0784  Mic   \n",
       "1       chr1     818121     818421       Opc_peak_3  5.43633     5.43633  Opc   \n",
       "2       chr1     827386     827686       Oli_peak_4  331.329     331.329  Oli   \n",
       "3       chr1     850353     850653       Exc_peak_4  4.58036     4.58036  Exc   \n",
       "4       chr1     869792     870092       Oli_peak_5  66.1156     66.1156  Oli   \n",
       "...      ...        ...        ...              ...      ...         ...  ...   \n",
       "406390  chrX  155820136  155820436  Oli_peak_112984  11.1663     11.1663  Oli   \n",
       "406391  chrX  155845361  155845661  Ast_peak_112992   5.6359      5.6359  Ast   \n",
       "406392  chrX  155874756  155875056  Exc_peak_156151   10.639      10.639  Exc   \n",
       "406393  chrX  155881107  155881407  Oli_peak_112985  413.806     413.806  Oli   \n",
       "406394  chrX  155956590  155956890  Ast_peak_112994   5.6359      5.6359  Ast   \n",
       "\n",
       "          rank  \n",
       "0        71146  \n",
       "1       365799  \n",
       "2        15933  \n",
       "3       262170  \n",
       "4        55696  \n",
       "...        ...  \n",
       "406390  176104  \n",
       "406391  241527  \n",
       "406392  179803  \n",
       "406393   13398  \n",
       "406394  241287  \n",
       "\n",
       "[406395 rows x 8 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = printer.uns['peak_calling']['merged']\n",
    "peaks = pd.DataFrame(df, columns=df.columns)\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f9670312-e0c8-4602-aadb-3d96689fa136",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks.to_csv(f'{work_dir}/tss6_atac_peaks.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "768afd2a-2344-402b-b6c0-12d264826057",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "continue\n",
      "continue\n",
      "continue\n",
      "continue\n",
      "continue\n",
      "continue\n",
      "continue\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 176485 peaks unresolved 104631 peaks selected\n",
      "round: 2 21244 peaks unresolved 10422 peaks selected\n",
      "round: 3 3508 peaks unresolved 1603 peaks selected\n",
      "round: 4 660 peaks unresolved 275 peaks selected\n",
      "round: 5 159 peaks unresolved 62 peaks selected\n",
      "round: 6 36 peaks unresolved 17 peaks selected\n",
      "round: 7 8 peaks unresolved 5 peaks selected\n",
      "round: 8 1 peaks unresolved 1 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 264613 peaks unresolved 156900 peaks selected\n",
      "round: 2 32718 peaks unresolved 16552 peaks selected\n",
      "round: 3 5259 peaks unresolved 2571 peaks selected\n",
      "round: 4 855 peaks unresolved 398 peaks selected\n",
      "round: 5 132 peaks unresolved 64 peaks selected\n",
      "round: 6 13 peaks unresolved 12 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 257232 peaks unresolved 154027 peaks selected\n",
      "round: 2 30197 peaks unresolved 15644 peaks selected\n",
      "round: 3 4579 peaks unresolved 2259 peaks selected\n",
      "round: 4 711 peaks unresolved 354 peaks selected\n",
      "round: 5 110 peaks unresolved 52 peaks selected\n",
      "round: 6 23 peaks unresolved 8 peaks selected\n",
      "round: 7 8 peaks unresolved 1 peaks selected\n",
      "round: 8 3 peaks unresolved 1 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 116986 peaks unresolved 73397 peaks selected\n",
      "round: 2 10911 peaks unresolved 6179 peaks selected\n",
      "round: 3 1222 peaks unresolved 648 peaks selected\n",
      "round: 4 167 peaks unresolved 84 peaks selected\n",
      "round: 5 36 peaks unresolved 14 peaks selected\n",
      "round: 6 7 peaks unresolved 3 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 252613 peaks unresolved 125716 peaks selected\n",
      "round: 2 50445 peaks unresolved 20613 peaks selected\n",
      "round: 3 11678 peaks unresolved 4720 peaks selected\n",
      "round: 4 2532 peaks unresolved 1048 peaks selected\n",
      "round: 5 509 peaks unresolved 223 peaks selected\n",
      "round: 6 103 peaks unresolved 50 peaks selected\n",
      "round: 7 19 peaks unresolved 7 peaks selected\n",
      "round: 8 6 peaks unresolved 3 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 126012 peaks unresolved 79443 peaks selected\n",
      "round: 2 12491 peaks unresolved 6785 peaks selected\n",
      "round: 3 1673 peaks unresolved 883 peaks selected\n",
      "round: 4 245 peaks unresolved 137 peaks selected\n",
      "round: 5 35 peaks unresolved 20 peaks selected\n",
      "round: 6 4 peaks unresolved 3 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 11401 peaks unresolved 10835 peaks selected\n",
      "round: 2 14 peaks unresolved 14 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n"
     ]
    }
   ],
   "source": [
    "scp.pp.call_peaks(\n",
    "    printer=printer,\n",
    "    frag_file=frags,\n",
    "    cell_grouping=cell_grouping,\n",
    "    group_names=list(group_names),\n",
    "    iterative_peak_merging=True,\n",
    "    merge_across_groups=False,\n",
    "    preset='chromvar',\n",
    "    sample_names=libs,\n",
    "    overwrite=False,\n",
    "    fdr_threshold=0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "972bbba5-9700-4a53-b75e-a71b43a77bd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellId</th>\n",
       "      <th>Individual</th>\n",
       "      <th>Batch</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>msex</th>\n",
       "      <th>ADdiag3types</th>\n",
       "      <th>nFrags</th>\n",
       "      <th>FRIP</th>\n",
       "      <th>ReadsInPeaks</th>\n",
       "      <th>TSSEnrichment</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>D19-8609_ATTGTCTTCTGGCTAA-1</th>\n",
       "      <td>D19-8609#ATTGTCTTCTGGCTAA-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>99891</td>\n",
       "      <td>0.325869</td>\n",
       "      <td>64964</td>\n",
       "      <td>6.236</td>\n",
       "      <td>D19-8609_ATTGTCTTCTGGCTAA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_TATGTTCTCTTCAGAG-1</th>\n",
       "      <td>D19-8609#TATGTTCTCTTCAGAG-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>98917</td>\n",
       "      <td>0.306965</td>\n",
       "      <td>60587</td>\n",
       "      <td>6.752</td>\n",
       "      <td>D19-8609_TATGTTCTCTTCAGAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_CATTCATTCGGAAAGC-1</th>\n",
       "      <td>D19-8609#CATTCATTCGGAAAGC-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Exc</td>\n",
       "      <td>C14</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>95683</td>\n",
       "      <td>0.312696</td>\n",
       "      <td>59730</td>\n",
       "      <td>6.747</td>\n",
       "      <td>D19-8609_CATTCATTCGGAAAGC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_TCAGGGCAGCACGTAG-1</th>\n",
       "      <td>D19-8609#TCAGGGCAGCACGTAG-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Inh</td>\n",
       "      <td>C20</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>94457</td>\n",
       "      <td>0.288590</td>\n",
       "      <td>54443</td>\n",
       "      <td>7.978</td>\n",
       "      <td>D19-8609_TCAGGGCAGCACGTAG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-8609_GTGTCCTAGCCTTTGA-1</th>\n",
       "      <td>D19-8609#GTGTCCTAGCCTTTGA-1</td>\n",
       "      <td>ROSMAP-33422</td>\n",
       "      <td>Library2</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C10</td>\n",
       "      <td>1</td>\n",
       "      <td>nonAD</td>\n",
       "      <td>93968</td>\n",
       "      <td>0.213309</td>\n",
       "      <td>39983</td>\n",
       "      <td>6.631</td>\n",
       "      <td>D19-8609_GTGTCCTAGCCTTTGA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_CGGACCACAGTCAGCC-1</th>\n",
       "      <td>D19-12527#CGGACCACAGTCAGCC-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1080</td>\n",
       "      <td>0.421296</td>\n",
       "      <td>910</td>\n",
       "      <td>11.279</td>\n",
       "      <td>D19-12527_CGGACCACAGTCAGCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_ACTTTCAGTCTAACCA-1</th>\n",
       "      <td>D19-12527#ACTTTCAGTCTAACCA-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1075</td>\n",
       "      <td>0.141860</td>\n",
       "      <td>305</td>\n",
       "      <td>6.178</td>\n",
       "      <td>D19-12527_ACTTTCAGTCTAACCA-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_GATCATGGTTTCCGGG-1</th>\n",
       "      <td>D19-12527#GATCATGGTTTCCGGG-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1062</td>\n",
       "      <td>0.322976</td>\n",
       "      <td>686</td>\n",
       "      <td>9.424</td>\n",
       "      <td>D19-12527_GATCATGGTTTCCGGG-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_ATGTCTTAGGGTTCCC-1</th>\n",
       "      <td>D19-12527#ATGTCTTAGGGTTCCC-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Oli</td>\n",
       "      <td>C3</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1033</td>\n",
       "      <td>0.296709</td>\n",
       "      <td>613</td>\n",
       "      <td>8.086</td>\n",
       "      <td>D19-12527_ATGTCTTAGGGTTCCC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D19-12527_TGGGTTACAATCACTG-1</th>\n",
       "      <td>D19-12527#TGGGTTACAATCACTG-1</td>\n",
       "      <td>ROSMAP-42915</td>\n",
       "      <td>Library7</td>\n",
       "      <td>Ast</td>\n",
       "      <td>C10</td>\n",
       "      <td>0</td>\n",
       "      <td>lateAD</td>\n",
       "      <td>1026</td>\n",
       "      <td>0.214425</td>\n",
       "      <td>440</td>\n",
       "      <td>8.628</td>\n",
       "      <td>D19-12527_TGGGTTACAATCACTG-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25949 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    cellId    Individual  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    D19-8609#ATTGTCTTCTGGCTAA-1  ROSMAP-33422   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    D19-8609#TATGTTCTCTTCAGAG-1  ROSMAP-33422   \n",
       "D19-8609_CATTCATTCGGAAAGC-1    D19-8609#CATTCATTCGGAAAGC-1  ROSMAP-33422   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    D19-8609#TCAGGGCAGCACGTAG-1  ROSMAP-33422   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    D19-8609#GTGTCCTAGCCTTTGA-1  ROSMAP-33422   \n",
       "...                                                    ...           ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1  D19-12527#CGGACCACAGTCAGCC-1  ROSMAP-42915   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  D19-12527#ACTTTCAGTCTAACCA-1  ROSMAP-42915   \n",
       "D19-12527_GATCATGGTTTCCGGG-1  D19-12527#GATCATGGTTTCCGGG-1  ROSMAP-42915   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  D19-12527#ATGTCTTAGGGTTCCC-1  ROSMAP-42915   \n",
       "D19-12527_TGGGTTACAATCACTG-1  D19-12527#TGGGTTACAATCACTG-1  ROSMAP-42915   \n",
       "\n",
       "                                 Batch Celltype Cluster  msex ADdiag3types  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_CATTCATTCGGAAAGC-1   Library2      Exc     C14     1        nonAD   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1   Library2      Inh     C20     1        nonAD   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1   Library2      Ast     C10     1        nonAD   \n",
       "...                                ...      ...     ...   ...          ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_GATCATGGTTTCCGGG-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  Library7      Oli      C3     0       lateAD   \n",
       "D19-12527_TGGGTTACAATCACTG-1  Library7      Ast     C10     0       lateAD   \n",
       "\n",
       "                              nFrags      FRIP  ReadsInPeaks  TSSEnrichment  \\\n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    99891  0.325869         64964          6.236   \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    98917  0.306965         60587          6.752   \n",
       "D19-8609_CATTCATTCGGAAAGC-1    95683  0.312696         59730          6.747   \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    94457  0.288590         54443          7.978   \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    93968  0.213309         39983          6.631   \n",
       "...                              ...       ...           ...            ...   \n",
       "D19-12527_CGGACCACAGTCAGCC-1    1080  0.421296           910         11.279   \n",
       "D19-12527_ACTTTCAGTCTAACCA-1    1075  0.141860           305          6.178   \n",
       "D19-12527_GATCATGGTTTCCGGG-1    1062  0.322976           686          9.424   \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1    1033  0.296709           613          8.086   \n",
       "D19-12527_TGGGTTACAATCACTG-1    1026  0.214425           440          8.628   \n",
       "\n",
       "                                                     index  \n",
       "D19-8609_ATTGTCTTCTGGCTAA-1    D19-8609_ATTGTCTTCTGGCTAA-1  \n",
       "D19-8609_TATGTTCTCTTCAGAG-1    D19-8609_TATGTTCTCTTCAGAG-1  \n",
       "D19-8609_CATTCATTCGGAAAGC-1    D19-8609_CATTCATTCGGAAAGC-1  \n",
       "D19-8609_TCAGGGCAGCACGTAG-1    D19-8609_TCAGGGCAGCACGTAG-1  \n",
       "D19-8609_GTGTCCTAGCCTTTGA-1    D19-8609_GTGTCCTAGCCTTTGA-1  \n",
       "...                                                    ...  \n",
       "D19-12527_CGGACCACAGTCAGCC-1  D19-12527_CGGACCACAGTCAGCC-1  \n",
       "D19-12527_ACTTTCAGTCTAACCA-1  D19-12527_ACTTTCAGTCTAACCA-1  \n",
       "D19-12527_GATCATGGTTTCCGGG-1  D19-12527_GATCATGGTTTCCGGG-1  \n",
       "D19-12527_ATGTCTTAGGGTTCCC-1  D19-12527_ATGTCTTAGGGTTCCC-1  \n",
       "D19-12527_TGGGTTACAATCACTG-1  D19-12527_TGGGTTACAATCACTG-1  \n",
       "\n",
       "[25949 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9c7a20ac-0344-4a77-8117-bdec3894763e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_grouping, group_names = scp.utils.df2cell_grouping(printer, meta[meta['ADdiag3types'] == 'nonAD'][['index', 'Celltype']])\n",
    "group_names = [f'CT_{x}' for x in group_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d2122115-0486-4368-b5f0-8ab36fa13616",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Mic_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Mic -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:37:44: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Mic_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Mic -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Mic\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Mic_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:44: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:44: #1 read treatment fragments... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Vas_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Vas -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:37:45: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Vas_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Vas -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Vas\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Vas_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:45: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:45: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:45:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #1 mean fragment size is determined as 165.1 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #1 fragment size = 165.1 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #1  total fragments in treatment: 638975 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:46:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Vas_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Vas_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:47:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:49:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:49: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Vas_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:49: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Vas_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:49: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Vas_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:49: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:50:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:51:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:52:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #1 mean fragment size is determined as 167.8 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #1 fragment size = 167.8 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #1  total fragments in treatment: 7433315 \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:37:57: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:01: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:01: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Mic_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:01: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Mic_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:01: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:01: #3 Call peaks for each chromosome... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Opc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Opc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:38:23: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Opc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Opc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Opc\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Opc_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:23: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:23: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:24:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:25:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:26:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:27:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:28: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Mic_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:28:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:28: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Mic_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:28: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Mic_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:28: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:29:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:30:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:31:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:32:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:33:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #1 mean fragment size is determined as 176.6 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #1 fragment size = 176.6 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #1  total fragments in treatment: 10848256 \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:40: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:45: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:45: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Opc_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:45: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Opc_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:45: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:38:45: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:18: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Opc_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:18: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Opc_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:18: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Opc_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:18: Done! \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Ast_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Ast -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:39:35: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Ast_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Ast -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Ast\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Ast_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:35: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:35: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:36:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:37:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:38:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:39:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:40:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:42:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:43:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:44:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:45:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:46:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:47:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:48:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:49:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:50:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:51:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:52:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:54:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:55:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:56:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:57:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:58:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:39:59:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:00:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:01:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:02:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:03:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:04:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:05:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:06:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #1 mean fragment size is determined as 170.2 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #1 fragment size = 170.2 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #1  total fragments in treatment: 29515342 \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:24: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:38: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:38: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Ast_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:38: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Ast_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:38: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:40:38: #3 Call peaks for each chromosome... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Inh_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Inh -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:41:34: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Inh_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Inh -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Inh\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Inh_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:34: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:34: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:35:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:35: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Ast_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:36: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Ast_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:36:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:36: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Ast_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:36: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:37:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:38:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:39:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:40:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:41:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:43:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:44:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:45:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:46:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:47:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:48:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:49:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:50:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:51:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:52:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:54:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:55:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:56:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:57:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:58:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:41:59:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:00:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:01:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:02:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:03:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:04:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:05:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:07:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:08:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:09:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:10:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:11:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:12:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:13:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:14:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:15:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:16:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:17:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:18:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:20:  42000000 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Exc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Exc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:42:41: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Exc_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Exc -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Exc\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Exc_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:41: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:41: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:42:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:43:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:44:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:45:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:46:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:47:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #1 mean fragment size is determined as 189.5 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #1 fragment size = 189.5 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #1  total fragments in treatment: 42675213 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:48:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:49:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:50:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:51:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:52:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:53:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:54:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:55:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:56:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:57:  16000000 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running macs2 with"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:42:58:  17000000 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " macs2 callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Oli_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Oli -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO  @ Sat, 15 Mar 2025 17:42:58: \n",
      "# Command line: callpeak --nomodel -t /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Oli_filtered_frag.tsv.gz --outdir /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2 -n CT_Oli -f BEDPE --nolambda --keep-dup all --call-summits --nomodel -B --SPMR --shift 75 --extsize 150 -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = CT_Oli\n",
      "# format = BEDPE\n",
      "# ChIP-seq file = ['/ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/CT_Oli_filtered_frag.tsv.gz']\n",
      "# control file = None\n",
      "# effective genome size = 2.70e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      "# Searching for subpeak summits is on\n",
      "# MACS will save fragment pileup signal per million reads\n",
      " \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:58: #1 read fragment files... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:58: #1 read treatment fragments... \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:59:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:42:59:  1000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:00:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:00:  2000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:01:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:02:  3000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:02:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:03:  4000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:03:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:04:  5000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:04:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:05:  6000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:05:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:06:  7000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:06:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:07:  8000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:07:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:08:  9000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:08:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Inh_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Inh_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:09:  10000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:10:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:10:  11000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:11:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:11:  12000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:12:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:12:  13000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:13:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:13:  14000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:14:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:15:  15000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:15:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:16:  16000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:16:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:17:  17000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:17:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:18:  18000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:18:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:19:  19000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:19:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:20:  20000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:20:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:21:  21000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:21:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:22:  22000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:22:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:23:  23000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:23:  42000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:24:  24000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:24:  43000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:25:  25000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:25:  44000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:26:  26000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:26:  45000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:27:  27000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:28:  46000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:28:  28000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:29:  47000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:30:  29000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:30:  48000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:31:  30000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:31:  49000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:32:  50000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:32:  31000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:33:  51000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:33:  32000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:34:  33000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:35:  34000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:36:  35000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:37:  36000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:38:  37000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:39:  38000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:40:  39000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:41:  40000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:42:  41000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:44:  42000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:45:  43000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:46:  44000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:47:  45000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:48:  46000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:49:  47000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:50:  48000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:51:  49000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:52:  50000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:53:  51000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:54:  52000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:55:  53000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:43:56:  54000000 \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #1 mean fragment size is determined as 192.7 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #1 fragment size = 192.7 \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #1  total fragments in treatment: 51357182 \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:05: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:29: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:29: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Exc_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:29: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Exc_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:29: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:29: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #1 mean fragment size is determined as 166.4 bp from treatment \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #1 fragment size = 166.4 \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #1  total fragments in treatment: 54617182 \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #1 finished! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #2 Build Peak Model... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #2 Skipped... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #3 Call peaks... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: # local lambda is disabled! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #3 !!!! DYNAMIC LAMBDA IS DISABLED !!!! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #3 Going to call summits inside each peak ... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:33: #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:35: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Inh_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:36: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Inh_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:36: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Inh_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:37: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:55: #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:55: #3   Write bedGraph files for treatment pileup (after scaling if necessary)... CT_Oli_treat_pileup.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:55: #3   Write bedGraph files for control lambda (after scaling if necessary)... CT_Oli_control_lambda.bdg \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:55: #3   --SPMR is requested, so pileup will be normalized by sequencing depth in million reads. \n",
      "INFO  @ Sat, 15 Mar 2025 17:44:55: #3 Call peaks for each chromosome... \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:07: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Exc_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:08: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Exc_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:08: #4 Write output xls file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Oli_peaks.xls \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:09: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Exc_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:09: #4 Write peak in narrowPeak format file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Oli_peaks.narrowPeak \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:09: Done! \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:09: #4 Write summits bed file... /ewsc/zhangruo/GAGE_seq_AD/scATAC/scprinter_tss6_supp/macs2/CT_Oli_summits.bed \n",
      "INFO  @ Sat, 15 Mar 2025 17:46:09: Done! \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 149405 peaks unresolved 92347 peaks selected\n",
      "round: 2 15704 peaks unresolved 7972 peaks selected\n",
      "round: 3 2382 peaks unresolved 1122 peaks selected\n",
      "round: 4 403 peaks unresolved 183 peaks selected\n",
      "round: 5 80 peaks unresolved 35 peaks selected\n",
      "round: 6 15 peaks unresolved 9 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 259907 peaks unresolved 154196 peaks selected\n",
      "round: 2 31586 peaks unresolved 16334 peaks selected\n",
      "round: 3 4875 peaks unresolved 2436 peaks selected\n",
      "round: 4 773 peaks unresolved 386 peaks selected\n",
      "round: 5 115 peaks unresolved 58 peaks selected\n",
      "round: 6 17 peaks unresolved 10 peaks selected\n",
      "round: 7 1 peaks unresolved 1 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 221451 peaks unresolved 137672 peaks selected\n",
      "round: 2 23078 peaks unresolved 12377 peaks selected\n",
      "round: 3 3224 peaks unresolved 1663 peaks selected\n",
      "round: 4 501 peaks unresolved 243 peaks selected\n",
      "round: 5 103 peaks unresolved 42 peaks selected\n",
      "round: 6 15 peaks unresolved 8 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 86219 peaks unresolved 59936 peaks selected\n",
      "round: 2 4913 peaks unresolved 3225 peaks selected\n",
      "round: 3 322 peaks unresolved 200 peaks selected\n",
      "round: 4 22 peaks unresolved 14 peaks selected\n",
      "round: 5 1 peaks unresolved 1 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 181021 peaks unresolved 96855 peaks selected\n",
      "round: 2 31537 peaks unresolved 13386 peaks selected\n",
      "round: 3 7043 peaks unresolved 2892 peaks selected\n",
      "round: 4 1539 peaks unresolved 658 peaks selected\n",
      "round: 5 287 peaks unresolved 131 peaks selected\n",
      "round: 6 45 peaks unresolved 20 peaks selected\n",
      "round: 7 6 peaks unresolved 3 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 101613 peaks unresolved 68169 peaks selected\n",
      "round: 2 7776 peaks unresolved 4548 peaks selected\n",
      "round: 3 828 peaks unresolved 455 peaks selected\n",
      "round: 4 94 peaks unresolved 65 peaks selected\n",
      "round: 5 9 peaks unresolved 5 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n",
      "Reading in peak summit file(s):\n",
      "NOTE: Assuming all start coordinates are 0-based ..\n",
      "\n",
      "Padding peak summits by: 400 bp on either side for\n",
      "Removing peaks overlapping with blacklisted regions and out of bound peaks based on chromosome sizes ..\n",
      "\n",
      "Filtering overlapping peaks based on peak summit score ..\n",
      "round: 1 3819 peaks unresolved 3731 peaks selected\n",
      "round: 2 2 peaks unresolved 2 peaks selected\n",
      "finish clearing\n",
      "finish sorting\n",
      "finished summary\n"
     ]
    }
   ],
   "source": [
    "scp.pp.call_peaks(\n",
    "    printer=printer,\n",
    "    frag_file=frags,\n",
    "    cell_grouping=cell_grouping,\n",
    "    group_names=list(group_names),\n",
    "    iterative_peak_merging=True,\n",
    "    merge_across_groups=False,\n",
    "    preset='chromvar',\n",
    "    sample_names=libs,\n",
    "    overwrite=False,\n",
    "    fdr_threshold=0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bacf70e-36df-4369-a24e-6ed25243e36a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e02a1024-47d0-4741-a8d4-202d0e096cd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:1: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  for key in printer.uns['peak_calling'].keys():\n",
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c7c2a18f21cc48f18556c95ac7b679a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/10849 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1000786e90594e39b3996700bf121825",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/80325 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3df9ee4d22b64696b7953521027982e8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/63376 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "080be8edbf39431387612cefc1b38fc1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/3733 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd424d879f994dce8de08607323427e2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/87271 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0b111119421f47719b3472255889e83c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/173421 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0814d116c80b4d1b8c99175f9070caa4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/101668 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0905478eb594c8aa6740bea35b0f0ed",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/152005 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1e487e24fbef4b708a75836512d55a80",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/152380 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93672e87a58746c2a80d98e4917df766",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/73242 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "093db9aa42a541aeaa8185c1974aa3ff",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/176497 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8e10fe432d142239c34461135ae7b68",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/172346 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9ea9f09b906c4d1eaf429a7d8d898e9f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/113945 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_401221/3531364185.py:3: DeprecationWarning: `Series._import_from_c` is deprecated. use _import_arrow_from_c; if you are using an extension, please compile it with latest 'pyo3-polars'\n",
      "  df = printer.uns['peak_calling'][key]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4bd3a6edd2714cba9b27978f3786008d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Making peak matrix:   0%|          | 0/117016 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for key in printer.uns['peak_calling'].keys():\n",
    "    if \"cleaned\" in key:\n",
    "        df = printer.uns['peak_calling'][key]\n",
    "        peaks = pd.DataFrame(df, columns=df.columns)\n",
    "        # peaks.to_csv(f'{work_dir}/tss6_{key}_atac_peaks.bed', sep='\\t', header=False, index=False)\n",
    "        atac_adata = scp.pp.make_peak_matrix(printer=printer,\n",
    "                                     regions=peaks,\n",
    "                                     region_width=300,\n",
    "                                     sparse=True)\n",
    "        atac_adata.obs = meta.loc[atac_adata.obs.index].copy()\n",
    "        atac_adata.write(f'{key}_cellxpeak.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b971895f-c46c-4184-bd82-a562ee6e0ad5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
