
############################
# sample meta-information -- manuscript
############################
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
library(ggbeeswarm)
# load the sample meta information
data_sample_meta <- read.table(file = "annotation/final_sample_meta_data.csv", sep = ',', header = T, stringsAsFactors = F, na.strings = 'NULL')
# plot each meta-data
data_sample_meta_filtered <- data_sample_meta %>%
  #filter(Dataset == "GAGE-seq") %>%
  filter(used_sample == 'Yes') %>%
  mutate(sample_id = paste0("P", projid)) %>%
  mutate(disease_group = ifelse(cogdx == 4, 'AD', 'CT')) %>%
  mutate(gender = ifelse(msex == 1, 'male', 'female')) %>%
  mutate(cogdx = factor(cogdx, levels = seq(1,6))) %>%
  mutate(braaksc = factor(braaksc, levels = seq(0,6))) %>%
  mutate(age_group = ifelse(age_death > 75 & age_death < 85, "75-85", ">85")) %>%
  mutate(gender_age_group = paste0(gender, '_', age_group)) %>%
  arrange(cogdx, braaksc, -cogn_global_lv)
# get the sample order 
sample_list_order <- unique(data_sample_meta_filtered$sample_id)
# Part-I: pathological
panel_meta_nft <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = nft)) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_distiller(palette = 'Reds', direction = 1) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
panel_meta_braaksc <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = factor(braaksc))) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  #scale_fill_brewer(palette = 'Reds', direction = 1) +
  scale_fill_manual(name = "braaksc", values = c('0' = "#d1e5f0", "1" = "#92c5de", "2" = "#4393c3", "3" = "#d9f0d3", "4" = "#7fbf7b", "5" = "#ef8a62", "6" = "#b2182b")) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
# Part-II: clinical
panel_meta_cognition <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = cogn_global_lv)) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_distiller(palette = 'Blues', direction = 1) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
panel_meta_cogdx <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = factor(cogdx))) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_manual(name = "cogdx", values = c("1" = "#4393c3", "4" = "#d6604d")) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
# Part-III: other information
panel_meta_pmi <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = pmi)) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_distiller(name = "pmi", palette = 'RdPu', direction = 1) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
panel_meta_gender <- data_sample_meta_filtered %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  ggplot(aes(x = sample_id, y = 1, fill = gender_age_group)) +
  geom_tile() +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_manual(name = "Gender", values = c("female_75-85" = "#c2a5cf", "female_>85" = "#9970ab", "male_75-85" = "#a6dba0", "male_>85" = "#5aae61")) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
# other features
panel_gage_amyloid <- data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = amyloid, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Amyloid level across regions") +
  #ggtitle("GAGE-seq data sample Amyloid level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
panel_gage_plaq_d <- data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = plaq_d, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Diffused plaque burden across regions") +
  #ggtitle("GAGE-seq data sample Diffused plaque level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
panel_gage_plaq_n <- data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = plaq_n, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Neuritic plaque burden across regions") +
  #ggtitle("GAGE-seq data sample Neuritic plaque burden level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
panel_gage_nft <-  data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = nft, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Neurofibrillary tangle burden across regions") +
  #ggtitle("GAGE-seq data sample Neurofibrillary tangle burden level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
panel_gage_tangle <-  data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = tangles, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Tangle density across regions") +
  #ggtitle("GAGE-seq data sample Neurofibrillary tangle burden level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
panel_gage_cogn <-  data_sample_meta_filtered %>%
  filter(Dataset == 'GAGE-seq') %>%
  ggplot(aes(x = disease_group, y = cogn_global_lv, color = disease_group)) +
  geom_quasirandom(width = 0.3) +
  xlab("") +
  ylab("Global cognition (z-score)") +
  #ggtitle("GAGE-seq data sample Neurofibrillary tangle burden level") +
  scale_color_manual(name = "", values = c("CT" = "#4393c3", "AD" = "#d6604d")) +
  coord_flip() +
  theme_cowplot()
# merge the plots - Figure 1
pdf(file = "R_notebook_figures/2025_Feb/figure_meta_sample.pdf", width = 7, height = 5)
plot_grid(panel_meta_nft, panel_meta_cognition, panel_meta_braaksc, panel_meta_cogdx, panel_meta_gender, panel_meta_pmi, ncol = 1, align = 'v', axis = 'lr')
dev.off()
# merge the plots - Supplement related to Figure 1
pdf(file = "R_notebook_figures/2025_Feb/figure_meta_sample_other_feature.pdf", width = 10, height = 6)
plot_grid(panel_gage_amyloid, panel_gage_plaq_d, panel_gage_plaq_n, panel_gage_nft, panel_gage_tangle, panel_gage_cogn, nrow = 2)
dev.off()
```


############################
# sample - major cell type fraction GAGE-seq & Xenium --manuscript
############################
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
library(beeswarm)
library(ggsignif)
# load the new cell annotation
data_cell_label <- read.table(file = "/work/magroup/shared/4DN_project/xinyue/spatial/meta-Seurat-spatial_Xenium_202412.csv", sep = ',', header = T, stringsAsFactors = F)
colnames(data_cell_label) <- c("cell_id", "sub")
# get the major
data_cell_label <- data_cell_label %>%
  mutate(major = gsub(" .*", "", sub))
data_cell_label_spatial <- data_cell_label %>%
  filter(!grepl("hPFC", cell_id)) %>%
  mutate(Dataset = "Xenium") %>%
  mutate(projid = gsub("_.*", "", cell_id))
data_cell_label_gageseq <- data_cell_label %>%
  filter(grepl("hPFC", cell_id)) %>%
  mutate(Dataset = "GAGE-seq")
# get gage-seq cell anno
data_cell_anno <- read.table(file = "results/202501/cell_meta_table.tsv", header = T, sep = '\t', stringsAsFactors = F)
data_cell_label_gageseq <- merge(data_cell_label_gageseq, data_cell_anno[, c("cell_id", "projid")], sort = F)
# merge with sample meta
data_cell_label_filtered <- rbind(data_cell_label_spatial, data_cell_label_gageseq)
# merge with sample meta (defined in the sample meta section)
data_cell_label_filtered <- merge(data_cell_label_filtered, data_sample_meta_filtered[, c("projid", "disease_group")], by = "projid", sort = F)
# sample_list_order is defined in sample meta-information section
panel_cell_type_fraction_major <- data_cell_label_filtered %>%
  mutate(sample_id = paste0("P", projid)) %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  filter(major %in% c("Exc", "Inh", "Astro", "Oligo", "OPC", "Micro", "Endo", "VLMC")) %>%
  mutate(major = factor(major, levels = c("Exc", "Inh", "Astro", "Oligo", "OPC", "Micro", "Endo", "VLMC"))) %>%
  group_by(Dataset, sample_id, major) %>%
  summarise(count = n()) %>%
  group_by(Dataset, sample_id) %>%
  mutate(total = sum(count)) %>%
  ggplot(aes(x = sample_id, y = count/total, fill = major))  +
  geom_col(color = "black") +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_manual(name = "", values = c("Exc" = "#30952c", "Inh" = "#a9da82", "Astro" = "#dd2124", "Oligo" = "#fe9b42", "OPC" = "#a7502a", "Micro" = "#c3a9cf", "Endo" = "#a6cee3", "VLMC" = "#1f78b4")) +
  scale_y_continuous(expand = c(0,0), label = scales::percent) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
# plot the fraction of sub-cell type
panel_cell_type_fraction_sub <- data_cell_label_filtered %>%
  mutate(sample_id = paste0("P", projid)) %>%
  mutate(sample_id = factor(sample_id, levels = sample_list_order)) %>%
  filter(major %in% c("Exc", "Inh", "Astro", "Oligo", "OPC", "Micro", "Endo", "VLMC")) %>%
  #mutate(sub = factor(sub, levels = c("Exc L2/3 IT", "Inh", "Astro", "Oligo", "OPC", "Micro", "Endo", "VLMC"))) %>%
  group_by(Dataset, sample_id, sub) %>%
  summarise(count = n()) %>%
  group_by(Dataset, sample_id) %>%
  mutate(total = sum(count)) %>%
  ggplot(aes(x = sample_id, y = count/total, fill = sub))  +
  geom_col(color = "black") +
  xlab("") +
  ylab("") +
  facet_grid(. ~ Dataset, scales = "free_x", drop = T, space = 'free') +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                                    "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                                    "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                                    "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                                    "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                                    "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                                    "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_y_continuous(expand = c(0,0), label = scales::percent) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = 'black')) +
  theme(strip.background = element_blank()) +
  theme(legend.key.size = unit(0.3, 'cm'))
# plot the 
panel_cell_type_fraction_major_test <- data_cell_label_filtered %>%
  filter(Dataset == "GAGE-seq") %>%
  group_by(disease_group, projid, major) %>%
  summarise(count = n()) %>%
  group_by(projid) %>%
  mutate(total = sum(count), major_fraction = count/total) %>%
  ggplot(aes(x = disease_group, y = major_fraction)) +
  geom_boxplot(outlier.shape = NA, aes(fill = major)) +
  geom_quasirandom(fill = "grey") +
  geom_signif(
    test = "wilcox.test",
    comparisons = list(c("AD", "CT")),
    map_signif_level = TRUE, textsize = 6
  ) +
  facet_wrap(~ major, scales = "free_y", nrow = 2) +
  scale_fill_manual(name = "", values = c("Exc" = "#30952c", "Inh" = "#a9da82", "Astro" = "#dd2124", "Oligo" = "#fe9b42", "OPC" = "#a7502a", "Micro" = "#c3a9cf", "Endo" = "#a6cee3", "VLMC" = "#1f78b4")) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.background = element_blank())
# merge the plots - figure 1
pdf(file = "R_notebook_figures/2025_Feb/figure_meta_sample.major_cell_fraction.pdf", width = 7, height = 5)
print(panel_cell_type_fraction_major)
dev.off()
# merge the plot - supp. figure 1
pdf(file = "R_notebook_figures/2025_Feb/figure_meta_sample.sub_cell_fraction.pdf", width = 10, height = 5)
print(panel_cell_type_fraction_sub)
dev.off()
pdf(file = "R_notebook_figures/2025_Feb/figure_meta_sample.major_cell_fraction.test.pdf", width =7, height = 5)
print(panel_cell_type_fraction_major_test)
dev.off()
```


############################
# integration snATAC-seq with GAGE-seq -- manuscript
############################
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
library(ggrastr)
# load the joint embedding results
data_atac_gageseq <- read.table(file = "/work/magroup/shared/4DN_project/ad/result_atac/seurat_obj/obj.int.atac.TSS6.umap_coordinates.csv", sep = ',', header = T, stringsAsFactors = F)
colnames(data_atac_gageseq)[1] <- "cell_id"
# plot the joint embedding UMAP between two techniques
panel_umap_atac_gageseq <- data_atac_gageseq %>%
  ggplot(aes(x = umap_1, y = umap_2, color = dataset)) +
  rasterise(geom_point(size = 0.1, alpha = 0.5), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("ATAC-seq" = "#fc8d62", "GAGE-seq" = "#8da0cb")) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) + 
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 1.0) +
  guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3))) 
# plot individual technique
panel_umap_atac <- data_atac_gageseq %>%
  filter(dataset == 'ATAC-seq') %>%
  ggplot(aes(x = umap_1, y = umap_2, color = cell_type)) +
  rasterise(geom_point(size = 0.1, alpha = 0.5), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("Ex" = "#30952c", "In" = "#a9da82", "Ast" = "#dd2124", "Oligo" = "#fe9b42", "Microglia" = "#c3a9cf", "OPC" = "#a7502a", "PerEndo" = "#a6cee3")) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) + 
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 1.0) +
  guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3))) 
panel_umap_gageseq <- data_atac_gageseq %>%
  filter(dataset == 'GAGE-seq') %>%
  ggplot(aes(x = umap_1, y = umap_2, color = cell_type)) +
  rasterise(geom_point(size = 0.1, alpha = 0.5), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("Exc" = "#30952c", "Inh" = "#a9da82", "Astro" = "#dd2124", "Oligo" = "#fe9b42", "Micro" = "#c3a9cf", "OPC" = "#a7502a", "Endo" = "#a6cee3", "VLMC" = "#1f78b4")) +
  theme_cowplot() +
  theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) + 
  theme(legend.position = "bottom") +
  theme(aspect.ratio = 1.0) +
  guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3))) 
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_integration_atac_gene_score_rna.v2.pdf", width = 15, height = 6)
plot_grid(panel_umap_atac_gageseq, panel_umap_atac, panel_umap_gageseq, nrow = 1, align = 'h', axis = 'tb')
dev.off()
```


############################
# integration PFC427 (matched 20 sample) with GAGE-seq -- manuscript
############################
```{r}
library(anndataR)
library(Seurat)
library(dplyr)
library(ggplot2)
library(cowplot)
library(ggrastr)
# load PFC427 data
obj.pfc427 <- read_h5ad("/work/magroup/shared/4DN_project/xinyue/reference/PFC427_raw_data_select_20_projids.h5ad")
# convert it to Seurat object
obj.pfc427 <- to_Seurat(obj.pfc427)
# update the metadata
Idents(obj.pfc427) <- obj.pfc427$major_cell_type
ad_projids <- NULL 
ct_projids <- NULL
# ad_projids and ct_projids have to be removed because the privacy issue
obj.pfc427$condition = ifelse(obj.pfc427$projid %in% ad_projids, "AD", 
                              ifelse(obj.pfc427$projid %in% ct_projids, "CT", "NA"))

# load GAGE-seq data
obj.gage <- readRDS("/work/magroup/shared/4DN_project/xinyue/RNA/202501/AD_rna_annotated.rds")
Idents(obj.gage) <- obj.gage@meta.data$major_cell_type

# integrate two techniques
obj.list <- list(obj.pfc427, obj.gage)
obj.list <- lapply(X=obj.list, FUN = function(x) {
    x <- NormalizeData(x, verbose = T)
    x <- FindVariableFeatures(x, verbose = T)
    x <- ScaleData(x, verbose = T)
})
# setting system parameters
print(obj.list)
options(future.globals.maxSize = 50000 * 1024^2)
library(future)
plan()
plan("multiprocess", workers = 48)
plan()
#
features <- SelectIntegrationFeatures(object.list = obj.list, verbose = T)
anchors <- FindIntegrationAnchors(object.list = obj.list, anchor.features = features, verbose = T)
saveRDS(anchors, file = "results/202501/integration_PFC427_GAGE_20_samples.anchors.rds")
anchors = readRDS(file = "results/202501/integration_PFC427_GAGE_20_samples.anchors.rds")
#
obj.int.pfc427 <- IntegrateData(anchorset = anchors, verbose = T)

# save integrated data
saveRDS(obj.int.pfc427, file = "results/202501/integration_PFC427_GAGE_20_sample.rds")

# load the integrated data
obj = readRDS("results/202501/integration_PFC427_GAGE_20_sample.rds")
num_pcs = 35
obj <- ScaleData(obj, verbose = T)
obj <- RunPCA(obj, npcs = num_pcs, verbose = T)
obj <- RunUMAP(obj, min.dist=0, n.neighbors=20, dims = 1:num_pcs, verbose=T)

# fix the dataset 
obj$dataset[is.na(obj$dataset)] = "PFC427_20sample"
#obj$seurat_clusters[obj$dataset == 'GAGE-seq'] = as.vector(Idents(obj.gage))

# extract the embedding
data_joint_embedding <- Embeddings(obj, reduction = "umap")
data_joint_embedding <- as.data.frame(data_joint_embedding)
data_joint_embedding$cell_id <- row.names(data_joint_embedding)
data_meta <- obj[[]]
data_meta <- data_meta[, c("major_cell_type", "projid", "dataset", "cell_type_high_resolution", "sub_cell_type")]
data_meta$sub_cell_type <- ifelse(is.na(data_meta$sub_cell_type), data_meta$cell_type_high_resolution, data_meta$sub_cell_type)
data_meta$cell_type_high_resolution <- NULL
data_meta$cell_id <- row.names(data_meta)
data_joint_embedding <- merge(data_joint_embedding, data_meta, by = "cell_id", sort = F)

# plotting
panel_joint_data_dataset <- data_joint_embedding %>%
  ggplot(aes(x = umap_1, y = umap_2, color = dataset)) +
  rasterise(geom_point(size = 0.4, alpha = 0.4), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("PFC427_20sample" = "#fc8d62", "GAGE-seq" = "#8da0cb")) +
  theme_cowplot() +
  theme(axis.ticks = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(axis.text = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3))) 
panel_joint_data_major_cell_type <- data_joint_embedding %>%
  filter(!is.na(major_cell_type)) %>%
  mutate(major_cell_type_dataset = paste0(dataset, '.', major_cell_type)) %>%
  #filter(major_cell_type_dataset %in% c("GAGE-seq.Endo", "PFC427_20sample.End", "GAGE-seq.VLMC", "PFC427_20sample.Fib", "PFC427_20sample.Per", "PFC427_20sample.T cells",  "PFC427_20sample.SMC", "PFC427_20sample.CAMs")) %>%
  ggplot(aes(x = umap_1, y = umap_2, color = factor(major_cell_type_dataset))) +
  rasterise(geom_point(size = 0.4, alpha = 0.5), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("GAGE-seq.Exc" = "#33a02c", "PFC427_20sample.Exc" = "#b2df8a",  "GAGE-seq.Inh" = "#1f78b4", "PFC427_20sample.Inh" = "#a6cee3",  "GAGE-seq.Astro" = "#e31a1c", "PFC427_20sample.Ast" = "#fb9a99", "GAGE-seq.Oligo" = "#ff7f00", "PFC427_20sample.Oli" = "#fdbf6f", "GAGE-seq.Micro" = "#6a3d9a", "PFC427_20sample.Mic" = "#cab2d6", "GAGE-seq.OPC" = "burlywood4", "PFC427_20sample.OPC" = "burlywood1", "GAGE-seq.Endo" = "deeppink", "PFC427_20sample.End" = "lightpink", "GAGE-seq.VLMC" = "gold3", "PFC427_20sample.Fib" = "khaki1", "PFC427_20sample.Per" = "khaki1", "PFC427_20sample.T cells" = "#cab2d6", "PFC427_20sample.SMC" = "khaki1", "PFC427_20sample.CAMs" = "#cab2d6")) +
  theme_cowplot() +
  theme(axis.ticks = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(axis.text = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(nrow = 6, byrow = TRUE, override.aes = list(size = 3))) 
panel_joint_data_sub_exc <- data_joint_embedding %>%
  filter(grepl("Exc", sub_cell_type)) %>%
  #filter(dataset == "GAGE-seq") %>%
  filter(!sub_cell_type %in% c("Exc NRGN", "Exc RELN CHD7")) %>%
  mutate(sub_cell_type_dataset = paste0(dataset, '.', sub_cell_type)) %>%
  ggplot(aes(x = umap_1, y = umap_2, color = factor(sub_cell_type_dataset))) +
  rasterise(geom_point(size = 0.4, alpha = 0.5), dpi = 300) +
  xlab("UMAP1") +
  ylab("UMAP2") +
  scale_color_manual(name = "", values = c("GAGE-seq.Exc L2/3 IT" = "#33a02c", "PFC427_20sample.Exc L2-3 CBLN2 LINC02306" = "#b2df8a",   "PFC427_20sample.Exc L3-4 RORB CUX2" = "#b2df8a", "GAGE-seq.Exc L4 IT" = "#1f78b4", "PFC427_20sample.Exc L3-5 RORB PLCH1" = "#a6cee3", "PFC427_20sample.Exc L4-5 RORB GABRG1" = "#a6cee3", "PFC427_20sample.Exc L4-5 RORB IL1RAPL2" = "#cab2d6", "GAGE-seq.Exc L5 IT" = "#6a3d9a", "PFC427_20sample.Exc L5-6 RORB LINC02196" = "#a6cee3", "GAGE-seq.Exc L5 ET" = "#e31a1c", "PFC427_20sample.Exc L5 ET" = "#fb9a99", "GAGE-seq.Exc L6 CT" = "#ff7f00", "PFC427_20sample.Exc L6 CT" = "#fdbf6f", "GAGE-seq.Exc L5/6 NP" = "cyan4", "PFC427_20sample.Exc L5/6 NP" = "cyan", "GAGE-seq.Exc L6 IT Car3" = "burlywood4", "PFC427_20sample.Exc L5/6 IT Car3" = "burlywood2", "GAGE-seq.Exc L6b" = "deeppink", "PFC427_20sample.Exc L6b" = "lightpink", "GAGE-seq.Exc L6 IT" = "gold3", "PFC427_20sample.Exc L6 THEMIS NFIA" = "khaki1")) +
  theme_cowplot() +
  theme(axis.ticks = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(axis.text = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(nrow = 8, byrow = TRUE, override.aes = list(size = 3))) 
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_integration_PFC427_20sample.pdf", width = 28, height = 12)
plot_grid(panel_joint_data_dataset, panel_joint_data_major_cell_type, panel_joint_data_sub_exc, nrow = 1, align = "h", axis = 'tb')
dev.off()

```


############################
# spatial & GAGE-seq joint embedding --- manuscript
############################
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
Sys.setenv(RETICULATE_PYTHON = "/home/yangz6/Software/miniconda3/envs/base2/envs/mc3c/bin/python")
library(reticulate)
library(anndataR)
library(ggplot2)
library(dplyr)
library(cowplot)
library(ggside)
library(ggrastr)
library(RColorBrewer)
# set conda
use_condaenv(condaenv = "mc3c", conda = "/home/yangz6/Software/miniconda3/envs/base2/bin/conda")
# load filtered spatial data
data_spatial <- read_h5ad("shared_xinyue/spatial/Xenium_12_2024_filtered_adata.h5ad")
data_spatial_obj <- data_spatial$to_Seurat()
# get spatial data meta
data_spatial_obs <- as.data.frame(data_spatial$obs)
colnames(data_spatial_obs) <- c("orig.ident", "nCount_RNA", "nFeature_RNA", "x_centroid", "y_centroid", "cell_area", "nucleus_area", "major", "projid", "disease_group")
data_spatial_obs$cell_id <- row.names(data_spatial_obs)
# update the new cell annotation
data_spatial_cell_label <- read.table(file = "/work/magroup/shared/4DN_project/xinyue/spatial/meta-Seurat-spatial_Xenium_202412.csv", sep = ',', header = T, stringsAsFactors = F)
colnames(data_spatial_cell_label) <- c("cell_id", "sub")
data_spatial_obs <- merge(data_spatial_obs, data_spatial_cell_label, by = "cell_id", sort = F)
# load spatial domain
data_spatial_domain <- read.table(file = "results/202501/spatial/Domain_label.csv", header = T, sep = ',', stringsAsFactors = F)
colnames(data_spatial_domain) <- c("cell_id", "domain_label")
data_spatial_domain <- data_spatial_domain %>%
  mutate(cell_id = gsub("-.$", "", cell_id))
data_spatial_obs <- merge(data_spatial_obs, data_spatial_domain, by = "cell_id", sort = F)
# load the umbedding data
data_joint_embedding <- read.table(file = "results/202501/spatial/umap_embeddings.tsv", header = T, sep = '\t', stringsAsFactors = F)
# load the GAGE-seq cell meta data
data_gageseq_obs <- read.table(file = "results/202501/cell_meta_table.tsv", header = T, sep = '\t', stringsAsFactors = F)
# extract embeddings for different modality separately
data_joint_embedding_spatial <- subset(data_joint_embedding, cell_id %in% data_spatial_obs$cell_id)
data_joint_embedding_spatial$gender <- "female"
data_joint_embedding_gageseq <- subset(data_joint_embedding, cell_id %in% data_gageseq_obs$cell_id)
data_joint_embedding_spatial <- merge(data_joint_embedding_spatial, data_spatial_obs, by = "cell_id", sort = F)
data_joint_embedding_gageseq <- merge(data_joint_embedding_gageseq, data_gageseq_obs, by = "cell_id", sort = F)

#######################
# plotting features on the embedding space
#######################
#
theme_spatial_embedding <- theme_cowplot() +
  theme(axis.text = element_blank()) +
  theme(axis.ticks = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(aspect.ratio = 1.0)
spectral_palette_4 <- colorRampPalette(brewer.pal(10, "Spectral"))(4)
spectral_palette_20 <- colorRampPalette(brewer.pal(10, "Spectral"))(20)
# Spatial
panel_embedding_spatial <- data_joint_embedding_spatial %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = sub)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                                    "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                                    "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                                    "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                                    "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                                    "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                                    "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_spatial_projid <- data_joint_embedding_spatial %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = factor(projid))) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Sample", values = spectral_palette_4) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_spatial_disease_group <- data_joint_embedding_spatial %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = disease_group)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Disease_group", values = c("AD" = "#f1a340", "CT" = "#998ec3")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_spatial_gender<- data_joint_embedding_spatial %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = gender)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Gender", values = c("male" = "#af8dc3", "female" = "#7fbf7b")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
# GAGE-seq
panel_embedding_gageseq <- data_joint_embedding_gageseq %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = sub)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                                    "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                                    "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                                    "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                                    "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                                    "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                                    "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_gageseq_projid <- data_joint_embedding_gageseq %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = factor(projid))) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Sample", values = spectral_palette_20) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_gageseq_disease_group <- data_joint_embedding_gageseq %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = disease_group)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Disease_group", values = c("AD" = "#f1a340", "CT" = "#998ec3")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
panel_embedding_gageseq_gender <- data_joint_embedding_gageseq %>%
  filter(!sub %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")) %>%
  ggplot(aes(x = rna_umap_1, y = rna_umap_2, color = gender)) +
  rasterise(geom_point(size = 0.1, alpha = 0.9), dpi = 300) +
  xlab("") +
  ylab("") +
  scale_color_manual(name = "Gender", values = c("male" = "#af8dc3", "female" = "#7fbf7b")) +
  theme_spatial_embedding +
  guides(color = guide_legend(override.aes = list(size = 2)))
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_spatial_integration_with_gage.sub.v2.pdf", width = 12, height = 36)
plot_grid(panel_embedding_spatial, panel_embedding_spatial_projid, panel_embedding_spatial_disease_group, panel_embedding_spatial_gender, ncol = 1, align = 'v', axis = 'lr')
plot_grid(panel_embedding_gageseq, panel_embedding_gageseq_projid, panel_embedding_gageseq_disease_group, panel_embedding_gageseq_gender, ncol = 1, align = 'v', axis = 'lr')
dev.off()
# domain label
cell_type_list <- sort(unique(data_joint_embedding_spatial$sub))
cell_type_list <- cell_type_list[!cell_type_list %in% c("Astro-Lowexpr", "Exc-Lowexpr", "Oligo-Lowexpr", "Micro-Lowexpr", "Unk")]
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_spatial_domain_level.split_view.pdf", width = 24, height = 5)
for (cell_type in cell_type_list) {
  tmp  <- data_joint_embedding_spatial %>%
    filter(sub == cell_type) %>%
    group_by(domain_label) %>%
    sample_frac(0.5) %>%
    mutate(domain_label = factor(domain_label, levels = c("L1", "L2", "L3", "L4", "L5", "L6", "WM")))
  plot1 <- ggplot() +
    rasterise(geom_point(data = tmp[, c("rna_umap_1", "rna_umap_2")], aes(x = rna_umap_1, y = rna_umap_2), size = 0.1, color = "lightgrey", alpha = 0.4), dpi = 200) +
    rasterise(geom_point(data = tmp, aes(x = rna_umap_1, y = rna_umap_2, color = domain_label), size = 0.1, alpha = 0.9), dpi = 200) +
    geom_density2d(data = tmp, aes(x = rna_umap_1, y = rna_umap_2), color = "black", alpha = 0.7) +
    xlab("") +
    ylab("") +
    ggtitle(paste0(cell_type, " (50% sampling)")) +
    #facet_grid(domain_label ~ sub, scales = "fixed") +
    facet_wrap(~ domain_label, scales = "fixed", nrow = 1, drop = F) +
    #scale_color_manual(name = "", values = c("L1" = "#66c2a5", "L2" = "#fc8d62", "L3" = "#8da0cb", "L4" = "#e78ac3", "L5" = "#a6d854", "L6" = "#ffd92f", "WM" = "#e5c494")) +
    scale_color_manual(name = "", values = c("L1" = "#99000d", "L2" = "#ef3b2c", "L3" = "#fc9272", "L4" = "#9ecae1", "L5" = "#4292c6", "L6" = "#08459f", "WM" = "#ffd92f")) +
    theme_spatial_embedding +
    guides(color = guide_legend(override.aes = list(size = 2))) +
    theme(strip.background = element_blank()) +
    theme(legend.position = 'none')
  cat(paste0(cell_type, ', '))
  print(plot1)
}
dev.off()

```


############################
# loop APA plot using mcool using 2025 loop -- manuscript
############################
```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
#########################
# APA plot
#########################
# load data
data_loop_apa_ad <- read.table(file= "results/202501/loop_pileup/major/AD_summary.tsv", header = T, sep = '\t', stringsAsFactors = F)
data_loop_apa_ct <- read.table(file = "results/202501/loop_pileup/major/CT_summary.tsv", header = T, sep = '\t', stringsAsFactors = F)
data_loop_apa_diff <- read.table(file = "results/202501/loop_pileup/major/AD_CT_diff_summary.tsv", header = T, sep = '\t', stringsAsFactors = F)
# add disease_group and merge
data_loop_apa_ad$disease_group = "AD"
data_loop_apa_ct$disease_group = "CT"
data_loop_apa <- merge(data_loop_apa_ad, data_loop_apa_ct, by = c("cell_type", "cluster_id", "ctcf_type", "loop_type", "row_idx", "col_idx"), sort = F, suffixes = c(".AD", ".CT"))
# plot cell type x loop type
panel_loop_pileup_strict <- data_loop_apa %>%
  filter(ctcf_type == "CTCF_strict") %>%
  mutate(loop_type = factor(loop_type, levels = c("whole", "whole_ctcf", "whole_other", "short_range_ctcf", "short_range_other", "mid_range_ctcf", "mid_range_other", "long_range_ctcf", "long_range_other"))) %>%
  mutate(ratio =  log2(value.AD) - log2(value.CT)) %>%
  ggplot(aes(y = row_idx, x = col_idx, fill = ratio)) +
  geom_tile() +
  facet_grid(cell_type + cluster_id ~ loop_type) +
  xlab("") +
  ylab("") +
  ggtitle("Raw loop pileup") +
  scale_fill_gradient2(low = "#0571b0", mid = "#f7f7f7", midpoint = 0, high = "#ca0020", limits = c(-0.1, 0.1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), trans = "reverse") +
  scale_x_continuous(expand = c(0,0)) +
  coord_fixed() +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(axis.ticks = element_blank(), axis.line = element_blank(), axis.text = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = "black")) +
  theme(strip.clip = 'off', strip.text.y = element_text(angle = 0, hjust = 0), strip.text.x = element_text(angle = 45, vjust = 0, hjust = 0))
# smoothed version
panel_loop_pileup_strict_smooth <- data_loop_apa_diff %>%
  filter(ctcf_type == "CTCF_strict") %>%
  mutate(loop_type = factor(loop_type, levels = c("whole", "whole_ctcf", "whole_other", "short_range_ctcf", "short_range_other", "mid_range_ctcf", "mid_range_other", "long_range_ctcf", "long_range_other"))) %>%
  mutate(ratio =  value.diff) %>%
  ggplot(aes(x = col_idx, y = row_idx, fill = ratio)) +
  geom_tile() +
  #geom_raster(interpolate = TRUE) +
  facet_grid(cell_type + cluster_id ~ loop_type) +
  xlab("") +
  ylab("") +
  ggtitle("Smothed loop pileup") +
  scale_fill_gradient2(low = "#0571b0", mid = "#f7f7f7", midpoint = 0, high = "#ca0020", limits = c(-0.08, 0.08), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), trans = "reverse") +
  scale_x_continuous(expand = c(0,0)) +
  coord_fixed() +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(axis.ticks = element_blank(), axis.line = element_blank(), axis.text = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = "black")) +
  theme(strip.clip = 'off', strip.text.y = element_text(angle = 0, hjust = 0), strip.text.x = element_text(angle = 45, vjust = 0, hjust = 0))
panel_loop_pileup_loose <- data_loop_apa %>%
  filter(ctcf_type == "CTCF_loose") %>%
  mutate(loop_type = factor(loop_type, levels = c("whole", "whole_ctcf", "whole_other", "short_range_ctcf", "short_range_other", "mid_range_ctcf", "mid_range_other", "long_range_ctcf", "long_range_other"))) %>%
  mutate(ratio =  log2(value.AD) - log2(value.CT)) %>%
  ggplot(aes(y = row_idx, x = col_idx, fill = ratio)) +
  geom_tile() +
  #geom_raster(interpolate = TRUE) +
  facet_grid(cell_type + cluster_id ~ loop_type) +
  xlab("") +
  ylab("") +
  ggtitle("Raw loop pileup") +
  scale_fill_gradient2(low = "#0571b0", mid = "#f7f7f7", midpoint = 0, high = "#ca0020", limits = c(-0.1, 0.1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), trans = "reverse") +
  scale_x_continuous(expand = c(0,0)) +
  coord_fixed() +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(axis.ticks = element_blank(), axis.line = element_blank(), axis.text = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = "black")) +
  theme(strip.clip = 'off', strip.text.y = element_text(angle = 0, hjust = 0), strip.text.x = element_text(angle = 45, vjust = 0, hjust = 0))
# smoothed version
panel_loop_pileup_loose_smooth <- data_loop_apa_diff %>%
  filter(ctcf_type == "CTCF_loose") %>%
  mutate(loop_type = factor(loop_type, levels = c("whole", "whole_ctcf", "whole_other", "short_range_ctcf", "short_range_other", "mid_range_ctcf", "mid_range_other", "long_range_ctcf", "long_range_other"))) %>%
  mutate(ratio =  value.diff) %>%
  ggplot(aes(x = col_idx, y = row_idx, fill = ratio)) +
  geom_tile() +
  facet_grid(cell_type + cluster_id ~ loop_type) +
  xlab("") +
  ylab("") +
  ggtitle("Smothed loop pileup") +
  scale_fill_gradient2(low = "#0571b0", mid = "#f7f7f7", midpoint = 0, high = "#ca0020", limits = c(-0.08, 0.08), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), trans = "reverse") +
  scale_x_continuous(expand = c(0,0)) +
  coord_fixed() +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(axis.ticks = element_blank(), axis.line = element_blank(), axis.text = element_blank()) +
  theme(panel.border = element_rect(linewidth = 1, color = "black")) +
  theme(strip.clip = 'off', strip.text.y = element_text(angle = 0, hjust = 0), strip.text.x = element_text(angle = 45, vjust = 0, hjust = 0))
pdf(file = "R_notebook_figures/2025_Feb/figure_loop_APA_plot.pdf", width = 10, height = 20)
print(panel_loop_pileup_strict)
print(panel_loop_pileup_strict_smooth)
print(panel_loop_pileup_loose)
print(panel_loop_pileup_loose_smooth)
dev.off()
```


############################
# long / short distribution  -- manuscript
############################
```{r}
library(ggplot2)
library(ggside)
library(dplyr)
library(tidyr)
library(cowplot)
library(circlize)
library(ComplexHeatmap)
library(ggsignif)
#######################
# function 
#######################
# load cell annotation
load_cell_annotation <- function(filename) {
  anno_cell <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
  # add major cell type,
  anno_cell <- anno_cell %>%
    mutate(sub2 = sub) %>%
    separate(sub2, sep = ' ', into = c("major", "Extra"), extra = "merge") %>% 
    select(-Extra)
  return(anno_cell)
}
# load sample annotation
load_sample_annotation <- function(filename) {
  anno_sample <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
}
# load distance track 
load_read_pair_distance_summary <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".pair_distance.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)
    # fix the typo in colnames
    names(table)[names(table) == "sampole_id"] <- "sample_id"
    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
#######################
# load data & annotation
#######################
# cell annotation
#anno_cell <- load_cell_annotation("results/202501/cell_meta_table.RNA.tsv")
anno_cell <- read.table(file = "results/202501/cell_meta_table.tsv", sep = '\t', header = T, stringsAsFactors = F)
# sample annotation
anno_sample <- load_sample_annotation("annotation/sample_info.v2.tsv")
# single-cell distance summary
patient_list_20 <- NULL 
# due to privary issue, the sample_ids are not provided here
data_pair_distance_mattew <- load_read_pair_distance_summary("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
#######################
# exploration 0: shrink the distance metric by aggregating bins 
#######################
# --- functions
get_distance_smooth <- function(data_pair_distance, agg_ratio = 4) {
  # construct cell by bin data.frame
  tmp <- data_pair_distance %>%
    # remove inter-chromosomal interaction
    filter(bin_idx != "inter") %>% 
    mutate(bin_idx = as.numeric(bin_idx)) %>%
    mutate(bin_idx = bin_idx %/% agg_ratio) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx) %>%
    mutate(bin_start = min(bin_start), bin_end = max(bin_end), cumu_count = max(cumu_count)) %>%
    select(cell_id, sample_id, distance_group, bin_idx, bin_start, bin_end, count, cumu_count) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx, bin_start, bin_end, cumu_count) %>%
    summarise(count = sum(count)) %>%
    group_by(cell_id, sample_id, distance_group) %>%
    mutate(total_count = sum(count)) %>%
    mutate(percent = count / total_count * 100)
  return(tmp)
}
get_distance_smooth_matrix <- function(data_pair_distance_smooth) {
  # build the percentage matrix
  tmp_matrix <- data_pair_distance_smooth %>% 
    ungroup() %>%
    group_by(cell_id) %>%
    select(cell_id, bin_idx, percent) %>%
    pivot_wider(names_from = "bin_idx", values_from = "percent")
  return(tmp_matrix)
}
get_cis_trans <- function(data_pair_distance) {
  # construct cell by bin data.frame
  tmp <- data_pair_distance %>%
    mutate(cis_trans = ifelse(bin_idx == 'inter', 'trans', 'cis')) %>%
    group_by(cell_id, sample_id, distance_group, cis_trans) %>%
    summarise(count = sum(count)) %>%
    group_by(cell_id, sample_id, distance_group) %>%
    mutate(total_count = sum(count)) %>%
    pivot_wider(names_from = "cis_trans", values_from = "count") %>%
    mutate(percent_cis = cis / total_count, percent_trans = trans / total_count, trans_cis_ratio = trans/cis)
  return(tmp)
}
# --- end
# build the distance percentage table - binning by four bins
data_distance_percent_smooth <- get_distance_smooth(data_pair_distance_mattew, agg_ratio = 4)
data_distance_percent_smooth_matrix <- get_distance_smooth_matrix(data_distance_percent_smooth)
# build the distance percentage table - raw binning
data_distance_percent_raw <- get_distance_smooth(data_pair_distance_mattew, agg_ratio = 1)
data_distance_percent_raw_matrix <- get_distance_smooth_matrix(data_distance_percent_raw)
# build the cis/trans summary
data_distance_cis_trans <- get_cis_trans(data_pair_distance_mattew)
#######################
# exploration I-A: erosion score, short-mid-long range
#######################
# --- functions -- 
get_distance_matrix_smooth_erosion <- function(data_distance_percent_smooth, data_distance_percent_smooth_matrix, sr_start, sr_end, lr_start, lr_end, agg_ratio = 4) {
  sr_start = sr_start * 4 / agg_ratio
  sr_end = sr_end * 4 / agg_ratio 
  lr_start = lr_start * 4 / agg_ratio
  lr_end = lr_end * 4 / agg_ratio 
  mid_start = sr_end + 1
  mid_end = lr_start - 1
  # get the bin_idx to bin range 
  bin_range <- unique(data_distance_percent_smooth[, c("bin_idx", "bin_start", "bin_end")])
  erosion_score_label_ls <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  erosion_score_label_lm <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", '(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')')
  erosion_score_label_ms <- paste0('(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')', " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  # get the erosion x cell id
  # note that the 1 here is used to compensate the first column cell_id
  data_erosion = data.frame(cell_id = data_distance_percent_smooth_matrix$cell_id, 
                            erosion_score_ls = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)]))), 
                            erosion_score_lm = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)]))), 
                            erosion_score_ms = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)]))), 
                            long_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])),
                            mid_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)])),
                            short_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)])))
  # scale the erosion score
  data_erosion$scaled_erosion_score_ls <- scale(data_erosion$erosion_score_ls, center = TRUE, scale = TRUE)
  data_erosion$scaled_erosion_score_lm <- scale(data_erosion$erosion_score_lm, center = TRUE, scale = TRUE)
  data_erosion$scaled_erosion_score_ms <- scale(data_erosion$erosion_score_ms, center = TRUE, scale = TRUE)
  # add erosion label
  data_erosion$erosion_label_ls <- erosion_score_label_ls
  data_erosion$erosion_label_lm <- erosion_score_label_lm
  data_erosion$erosion_label_ms <- erosion_score_label_ms
  return(data_erosion)
}
get_distance_matrix_raw_erosion <- function(data_distance_percent_smooth, data_distance_percent_smooth_matrix, sr_start, sr_end, lr_start, lr_end) {
  # get the bin_idx to bin range 
  mid_start = sr_end + 1
  mid_end = lr_start - 1
  bin_range <- unique(data_distance_percent_smooth[, c("bin_idx", "bin_start", "bin_end")])
  erosion_score_label_ls <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  erosion_score_label_lm <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", '(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')')
  erosion_score_label_ms <- paste0('(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')', " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  # get the erosion x cell id
  # note that the 1 here is used to compensate the first column cell_id
  data_erosion = data.frame(cell_id = data_distance_percent_smooth_matrix$cell_id, 
                            erosion_score_ls = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)]))), 
                            erosion_score_lm = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)]))), 
                            erosion_score_ms = log2( rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)])) / rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)]))), 
                            long_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (lr_start+1):(lr_end+1)])),
                            mid_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (mid_start+1):(mid_end+1)])),
                            short_range_frac = rowSums(as.matrix(data_distance_percent_smooth_matrix[, (sr_start+1):(sr_end+1)])))
  # scale the erosion score
  data_erosion$scaled_erosion_score_ls <- scale(data_erosion$erosion_score_ls, center = TRUE, scale = TRUE)
  data_erosion$scaled_erosion_score_lm <- scale(data_erosion$erosion_score_lm, center = TRUE, scale = TRUE)
  data_erosion$scaled_erosion_score_ms <- scale(data_erosion$erosion_score_ms, center = TRUE, scale = TRUE)
  # add erosion label
  data_erosion$erosion_label_ls <- erosion_score_label_ls
  data_erosion$erosion_label_lm <- erosion_score_label_lm
  data_erosion$erosion_label_ms <- erosion_score_label_ms
  return(data_erosion)
}
# --- end
data_distance_erosion_smooth <- get_distance_matrix_smooth_erosion(data_distance_percent_smooth, data_distance_percent_smooth_matrix, 1, 14, 25, 32, agg_ratio = 4)
data_distance_erosion_raw <- get_distance_matrix_raw_erosion(data_distance_percent_raw, data_distance_percent_raw_matrix, 1, 56, 97, 128)
# merge with cell annotation
data_distance_erosion_smooth_cell <- merge(data_distance_erosion_smooth, anno_cell, by = "cell_id")
data_distance_erosion_raw_cell <- merge(data_distance_erosion_raw, anno_cell, by = "cell_id")
# merge with cis/trans information
data_distance_erosion_smooth_cell <- merge(data_distance_erosion_smooth_cell, data_distance_cis_trans[, c("cell_id", "total_count", "cis", "trans", "percent_cis", "percent_trans", "trans_cis_ratio")], by = "cell_id", sort = F)
data_distance_erosion_raw_cell <- merge(data_distance_erosion_raw_cell, data_distance_cis_trans[, c("cell_id", "total_count", "cis", "trans", "percent_cis", "percent_trans", "trans_cis_ratio")], by = "cell_id", sort = F)
# save the results to files
write.table(data_distance_erosion_smooth_cell, file = "results/202501/cool_file/distance_erosion_smooth.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
write.table(data_distance_erosion_raw_cell, file = "results/202501/cool_file/distance_erosion_raw.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
#data_distance_erorion_smooth_cell <- read.table(file = "results/202501/cool_file/distance_erosion_smooth.tsv", sep = '\t', header = T, stringsAsFactors = F)
#data_distance_erosion_raw_cell <- read.table(file = "results/202501/cool_file/distance_erosion_raw.tsv", sep = '\t', header = T, stringsAsFactors = F)
########################
# exploration II-A: distance vs contact frequency cluster heatmap (raw)
########################
# step 1: get the overall long vs short range regardless of disease_group
global_erosion_score_rank_major <- data_distance_erosion_raw_cell %>%
  group_by(major) %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
global_erosion_score_rank_sub <- data_distance_erosion_raw_cell %>%
  group_by(sub) %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
# step 2: for AD and CT rank the cells by global rank major then by global rank sub
global_raw_erosion_ad <- data_distance_erosion_raw_cell %>%
  filter(disease_group == 'AD') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(major, sub, erosion_score_ls)
global_raw_erosion_ct <- data_distance_erosion_raw_cell %>%
  filter(disease_group == 'CT') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(major, sub, erosion_score_ls)
# step 3: convert the data.frame to matrix, then use complexheatmap to plot the heatmap
# shows the total count, cis-trans ratio, gender as column annotation
# ----- AD -----
global_raw_erosion_mat_ad <- data_distance_percent_raw_matrix %>%
  ungroup() %>%
  filter(cell_id %in% global_raw_erosion_ad$cell_id) %>%
  mutate(cell_id = factor(cell_id, levels = global_raw_erosion_ad$cell_id)) %>%
  arrange(cell_id)
global_raw_erosion_mat_ad.mat <- as.matrix(global_raw_erosion_mat_ad[, 2:ncol(global_raw_erosion_mat_ad)])
rownames(global_raw_erosion_mat_ad.mat) <- global_raw_erosion_mat_ad$cell_id
global_raw_erosion_mat_ad <- t(global_raw_erosion_mat_ad.mat)
global_raw_erosion_mat_ad <- global_raw_erosion_mat_ad[nrow(global_raw_erosion_mat_ad):1, ]
rm(global_raw_erosion_mat_ad.mat)
# add columns' annotation
column_ha_ad = HeatmapAnnotation(sub = global_raw_erosion_ad$sub,
                                 major = global_raw_erosion_ad$major,
                                 zscore_long_short = global_raw_erosion_ad$scaled_erosion_score_ls,
                                 #zscore_long_mid = global_raw_erosion_ad$scaled_erosion_score_lm,
                                 #zscore_mid_short = global_raw_erosion_ad$scaled_erosion_score_ms,
                                 log2_trans_cis = log2(global_raw_erosion_ad$trans_cis_ratio), 
                                 #log10_read_pair_count = anno_barplot(log10(global_raw_erosion_ad$total_count)),
                                 show_annotation_name = FALSE,
                                 col = list(sub = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                                    "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                                    "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                                    "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                                    "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                                    "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                                    "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae"),
                                            major = c("Astro" = "#665d48bb", "Endo" = "#8e6c62bb", "Exc" = "#b4d335bb", "Inh" = "#f8981dbb", "Micro" = "#95ae96bb", "Oligo" = "#53776cbb", "OPC" = "#384a46bb", "VLMC" = "#6b7257bb"),
                                            zscore_long_short = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            #zscore_long_mid = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            #zscore_mid_short = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            log2_trans_cis = colorRamp2(c(-2, 0), hcl_palette = "Viridis")
                                 )
)
ht_ad <- Heatmap(global_raw_erosion_mat_ad, name = "percent_ad", column_title = "Read pair distance plot (AD)",
                 cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = F, show_column_names = F,
                 col =  colorRamp2(c(0.05, 1.0, 1.8), c("white", "#6baed6", "#08519c")),
                 use_raster = T, border_gp = gpar(col = "black", lty = 1), 
                 bottom_annotation = column_ha_ad)
# ----- CT -----
global_raw_erosion_mat_ct <- data_distance_percent_raw_matrix %>%
  ungroup() %>%
  filter(cell_id %in% global_raw_erosion_ct$cell_id) %>%
  mutate(cell_id = factor(cell_id, levels = global_raw_erosion_ct$cell_id)) %>%
  arrange(cell_id)
global_raw_erosion_mat_ct.mat <- as.matrix(global_raw_erosion_mat_ct[, 2:ncol(global_raw_erosion_mat_ct)])
rownames(global_raw_erosion_mat_ct.mat) <- global_raw_erosion_mat_ct$cell_id
global_raw_erosion_mat_ct <- t(global_raw_erosion_mat_ct.mat)
global_raw_erosion_mat_ct <- global_raw_erosion_mat_ct[nrow(global_raw_erosion_mat_ct):1, ]
rm(global_raw_erosion_mat_ct.mat)
# add columns' annotation
column_ha_ct = HeatmapAnnotation(sub = global_raw_erosion_ct$sub,
                                 major = global_raw_erosion_ct$major,
                                 zscore_long_short = global_raw_erosion_ct$scaled_erosion_score_ls,
                                 #zscore_long_mid = global_raw_erosion_ct$scaled_erosion_score_lm,
                                 #zscore_mid_short = global_raw_erosion_ct$scaled_erosion_score_ms,
                                 log2_trans_cis = log2(global_raw_erosion_ct$trans_cis_ratio), 
                                 #log10_read_pair_count = anno_barplot(log10(global_raw_erosion_ct$total_count)),
                                 show_annotation_name = FALSE,
                                 col = list(sub = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                                    "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                                    "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                                    "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                                    "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                                    "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                                    "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae"),
                                            major = c("Astro" = "#665d48bb", "Endo" = "#8e6c62bb", "Exc" = "#b4d335bb", "Inh" = "#f8981dbb", "Micro" = "#95ae96bb", "Oligo" = "#53776cbb", "OPC" = "#384a46bb", "VLMC" = "#6b7257bb"),
                                            zscore_long_short = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            #zscore_long_mid = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            #zscore_mid_short = colorRamp2(c(-2, 0, 2), hcl_palette = "RdBu", reverse = TRUE),
                                            log2_trans_cis = colorRamp2(c(-2, 0), hcl_palette = "Viridis")
                                 )
)
ht_ct <- Heatmap(global_raw_erosion_mat_ct, name = "percent_ct", column_title = "Read pair distance plot (CT)",
                 cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = F, show_column_names = F,
                 col =  colorRamp2(c(0.05, 1.0, 1.8), c("white", "#6baed6", "#08519c")),
                 use_raster = T, border_gp = gpar(col = "black", lty = 1), 
                 bottom_annotation = column_ha_ct)
ht_distance_range <- rowAnnotation(group = c(rep('other', 15), rep('long', 43), rep('mid', 29), rep('short', 57)), show_annotation_name = FALSE,
                                   col = list (group = c("other" = "white", "long" = "grey35", "mid" = "grey60", "short" = "grey85")))
# step 4: concat the heatmap
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_heatmap.pdf", width = 20, height = 7)
ht_list = ht_ad + ht_ct + ht_distance_range
draw(ht_list)
dev.off()

########################
# exploration II-B: distance vs contact frequency cluster boxplot (sub-cell type)
########################
# group by disease_group, then by sub-cell type
calculate_boxplot_pvalue <- function(input_data, colname) {
  result_summary = data.frame()
  for (cell_type in unique(input_data$sub)) {
    AD_group = subset(input_data, sub == cell_type & disease_group == "AD")[, colname]
    CT_group = subset(input_data, sub == cell_type & disease_group == "CT")[, colname]
    result <- wilcox.test(unlist(AD_group), unlist(CT_group), paired = F, alternative = "two.sided")
    data <- data.frame(cell_type = cell_type, p_value = result$p.value, AD_cell_count = length(AD_group), CT_cell_count = length(CT_group), test_feature = colname)
    result_summary <- rbind(result_summary, data)
  }
  return(result_summary)
}
#
test_global_erosion_ls_boxplot_cell <- calculate_boxplot_pvalue(data_distance_erosion_raw_cell, "scaled_erosion_score_ls")
panel_global_erosion_ls_boxplot_cell <- data_distance_erosion_raw_cell %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = scaled_erosion_score_ls, fill = sub, alpha = disease_group)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("log2 long vs short interaction") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
test_global_erosion_lm_boxplot_cell <- calculate_boxplot_pvalue(data_distance_erosion_raw_cell, "scaled_erosion_score_lm")
panel_global_erosion_lm_boxplot_cell <- data_distance_erosion_raw_cell %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = scaled_erosion_score_lm, fill = sub, alpha = disease_group)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("log2 long vs mid interaction") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
test_global_erosion_ms_boxplot_cell <- calculate_boxplot_pvalue(data_distance_erosion_raw_cell, "scaled_erosion_score_ms")
panel_global_erosion_ms_boxplot_cell <- data_distance_erosion_raw_cell %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = scaled_erosion_score_ms, fill = sub, alpha = disease_group)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("log2 mid vs short interaction") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
test_global_trans_cis_boxplot_cell <- calculate_boxplot_pvalue(data_distance_erosion_raw_cell, "trans_cis_ratio")
panel_global_trans_cis_boxplot_cell <- data_distance_erosion_raw_cell %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = trans_cis_ratio, fill = sub, alpha = disease_group)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("cis/trans ratio") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "none")
## test patient
data_distance_erosion_raw_cell <- merge(data_distance_erosion_raw_cell, anno_sample[, c("projid", "age_group")], by = "projid", sort = F)
data_distance_erosion_raw_cell <- data_distance_erosion_raw_cell %>%
  mutate(age_group_new = ifelse(age_group %in% c("75-80", "80-85"), "75-85", "85+"))
# group by disease_group, then by sub-cell type
calculate_boxplot_pvalue_age_group <- function(input_data, colname) {
  result_summary = data.frame()
  for (cell_type in unique(input_data$sub)) {
    AD_group_1 = subset(input_data, sub == cell_type & disease_group == "AD" & age_group_new == "75-85")[, colname]
    AD_group_2 = subset(input_data, sub == cell_type & disease_group == "AD" & age_group_new == "85+")[, colname]
    CT_group_1 = subset(input_data, sub == cell_type & disease_group == "CT" & age_group_new == "75-85")[, colname]
    CT_group_2 = subset(input_data, sub == cell_type & disease_group == "CT" & age_group_new == "85+")[, colname]
    #
    result <- tryCatch(
      {
        # Attempt Wilcoxon test
        test <- wilcox.test(AD_group_1, AD_group_2, paired = FALSE, alternative = "two.sided")
        # Return results as a structured list
        list(p.value = test$p.value, statistic = test$statistic, error = NULL)
      },
      error = function(e) { list(p.value = NA, statistic = NA,error = e$message)}
    )
    data <- data.frame(cell_type = cell_type, p_value = result$p.value, cell_count_1 = length(AD_group_1), cell_count_2 = length(AD_group_2), test_feature = colname, group = "AD_young_vs_old")
    result_summary <- rbind(result_summary, data)
    #
    result <- tryCatch(
      {
        # Attempt Wilcoxon test
        test <- wilcox.test(CT_group_1, CT_group_2, paired = FALSE, alternative = "two.sided")
        # Return results as a structured list
        list(p.value = test$p.value, statistic = test$statistic, error = NULL)
      },
      error = function(e) { list(p.value = NA, statistic = NA,error = e$message)}
    )
    data <- data.frame(cell_type = cell_type, p_value = result$p.value, cell_count_1 = length(CT_group_1), cell_count_2 = length(CT_group_2), test_feature = colname, group = "CT_young_vs_old")
    result_summary <- rbind(result_summary, data)
    #
    result <- tryCatch(
      {
        # Attempt Wilcoxon test
        test <- wilcox.test(AD_group_1, CT_group_1, paired = FALSE, alternative = "two.sided")
        # Return results as a structured list
        list(p.value = test$p.value, statistic = test$statistic, error = NULL)
      },
      error = function(e) { list(p.value = NA, statistic = NA,error = e$message)}
    )
    data <- data.frame(cell_type = cell_type, p_value = result$p.value, cell_count_1 = length(AD_group_1), cell_count_2 = length(CT_group_1), test_feature = colname, group = "young_AD_vs_CT")
    result_summary <- rbind(result_summary, data)
    #
    result <- tryCatch(
      {
        # Attempt Wilcoxon test
        test <- wilcox.test(AD_group_2, CT_group_2, paired = FALSE, alternative = "two.sided")
        # Return results as a structured list
        list(p.value = test$p.value, statistic = test$statistic, error = NULL)
      },
      error = function(e) { list(p.value = NA, statistic = NA,error = e$message)}
    )
    data <- data.frame(cell_type = cell_type, p_value = result$p.value, cell_count_1 = length(AD_group_2), cell_count_2 = length(CT_group_2), test_feature = colname, group = "old_AD_vs_CT")
    result_summary <- rbind(result_summary, data)
    cat(paste0("process ", cell_type, " done", '\n'))
  }
  return(result_summary)
}
test_global_erosion_age_group <- calculate_boxplot_pvalue_age_group(data_distance_erosion_raw_cell, "scaled_erosion_score_ls")
#
panel_global_erosion_ls_boxplot_age_group <- data_distance_erosion_raw_cell %>%
  #filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = scaled_erosion_score_ls, fill = sub, alpha = disease_group)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("log2 long vs short interaction") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2",
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9",
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e",
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037",
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ age_group_new, nrow = 2) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  #theme(legend.position = "none") +
  theme(strip.background = element_blank())
panel_global_erosion_ls_boxplot_age_group_disease_group <- data_distance_erosion_raw_cell %>%
  #filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = sub, y = scaled_erosion_score_ls, fill = sub, alpha = age_group_new)) +
  geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
  xlab("") +
  ylab("log2 long vs short interaction") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2",
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9",
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e",
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037",
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  #scale_color_manual(name = "", values = c("75-85" = "darkgrey", "85+" = "black")) +
  facet_wrap(~ disease_group, nrow = 2) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  #theme(legend.position = "none") +
  theme(strip.background = element_blank())
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.agegroup.pdf", height = 6, width = 12)
print(panel_global_erosion_ls_boxplot_age_group)
print(panel_global_erosion_ls_boxplot_age_group_disease_group)
dev.off()
#
data_distance_erosion_raw_patient <-  data_distance_erosion_raw_cell %>%
  group_by(disease_group, projid, gender, sub) %>%
  summarise(scaled_erosion_score_ls_ave = mean(scaled_erosion_score_ls, na.rm = T),
            scaled_erosion_score_lm_ave = mean(scaled_erosion_score_lm, na.rm = T),
            scaled_erosion_score_ms_ave = mean(scaled_erosion_score_ms, na.rm = T),
            log2_trans_cis_ratio = mean(log2(trans_cis_ratio))) 
test_global_erosion_ls_boxplot_patient <- calculate_boxplot_pvalue(data_distance_erosion_raw_patient, "scaled_erosion_score_ls_ave")
test_global_erosion_lm_boxplot_patient <- calculate_boxplot_pvalue(data_distance_erosion_raw_patient, "scaled_erosion_score_lm_ave")
test_global_erosion_ms_boxplot_patient <- calculate_boxplot_pvalue(data_distance_erosion_raw_patient, "scaled_erosion_score_ms_ave")
#
panel_global_erosion_ls_boxplot_patient <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(disease_group, projid, gender, sub) %>%
  summarise(erosion_score_ave = mean(scaled_erosion_score_ls),
            log2_trans_cis_ratio = mean(log2(trans_cis_ratio))) %>%
  ggplot(aes(x = disease_group, y = erosion_score_ave, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("log2 long vs short interaction") +
  facet_wrap(~ sub, nrow = 1) +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_lm_boxplot_patient <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(disease_group, projid, gender, sub) %>%
  summarise(erosion_score_ave = mean(scaled_erosion_score_lm),
            log2_trans_cis_ratio = mean(log2(trans_cis_ratio))) %>%
  ggplot(aes(x = disease_group, y = erosion_score_ave, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("log2 long vs mid interaction") +
  facet_wrap(~ sub, nrow = 1) +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_ms_boxplot_patient <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(disease_group, projid, gender, sub) %>%
  summarise(erosion_score_ave = mean(scaled_erosion_score_ms),
            log2_trans_cis_ratio = mean(log2(trans_cis_ratio))) %>%
  ggplot(aes(x = disease_group, y = erosion_score_ave, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("log2 mid vs short interaction") +
  facet_wrap(~ sub, nrow = 1) +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.pdf", height = 28, width = 9)
plot_grid(panel_global_erosion_ls_boxplot_cell, panel_global_erosion_lm_boxplot_cell, panel_global_erosion_ms_boxplot_cell, panel_global_trans_cis_boxplot_cell, panel_global_erosion_ls_boxplot_patient, panel_global_erosion_lm_boxplot_patient, panel_global_erosion_ms_boxplot_patient, ncol = 1)
dev.off()
# save all test
write.table(test_global_erosion_age_group, file= "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.age_group.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
test_global_erosion <- rbind(test_global_erosion_ls_boxplot_cell, test_global_erosion_lm_boxplot_cell, test_global_erosion_ms_boxplot_cell, test_global_trans_cis_boxplot_cell)
write.table(test_global_erosion, file= "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
# pairwise scatterplot
panel_global_trans_cis_vs_erosion_ls <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = scale(log2(trans_cis_ratio)), y = scaled_erosion_score_ls, color = sub)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  geom_point(size = 0.5, alpha = 0.6) +
  facet_grid(disease_group ~ sub) +
  xlab("zscore - log2 trans/cis ratio") +
  ylab("zscore - log2 long/short interaction ratio") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 45, hjust = 0, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none")
panel_global_trans_cis_vs_erosion_lm <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = scale(log2(trans_cis_ratio)), y = scaled_erosion_score_lm, color = sub)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  geom_point(size = 0.5, alpha = 0.6) +
  facet_grid(disease_group ~ sub) +
  xlab("zscore - log2 trans/cis ratio") +
  ylab("zscore - log2 long/mid interation ratio") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 45, hjust = 0, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none")
panel_global_trans_cis_vs_erosion_ms <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = scale(log2(trans_cis_ratio)), y = scaled_erosion_score_ms, color = sub)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  geom_point(size = 0.5, alpha = 0.6) +
  facet_grid(disease_group ~ sub) +
  xlab("zscore - log2 trans/cis ratio") +
  ylab("zscore - log2 mid/short interaction ratio") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 45, hjust = 0, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none")
panel_global_erosion_ls_vs_erosion_lm <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = scaled_erosion_score_ls, y = scaled_erosion_score_lm, color = sub)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  geom_point(size = 0.5, alpha = 0.6) +
  facet_grid(disease_group ~ sub) +
  xlab("zscore - log2 long/short interaction ratio") +
  ylab("zscore - log2 long/mid interaction ratio") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 45, hjust = 0, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none")
panel_global_erosion_ls_vs_erosion_ms <- data_distance_erosion_raw_cell %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = scaled_erosion_score_ls, y = scaled_erosion_score_ms, color = sub)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  geom_point(size = 0.5, alpha = 0.6) +
  facet_grid(disease_group ~ sub) +
  xlab("zscore - log2 long/short interaction ratio") +
  ylab("zscore - log2 mid/short interaction ratio") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 45, hjust = 0, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none")
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_scatter.pdf", height = 5, width =36)
print(panel_global_trans_cis_vs_erosion_ls)
print(panel_global_trans_cis_vs_erosion_lm)
print(panel_global_trans_cis_vs_erosion_ms)
print(panel_global_erosion_ls_vs_erosion_lm)
print(panel_global_erosion_ls_vs_erosion_ms)
dev.off()
########################
# exploration II-C: distance vs gene expression (global)
#########################
# merge
data_distance_erosion_raw_cell_rna <- merge(data_distance_erosion_raw_cell, data_rna_meta, by = "cell_id", suffix = c(".dna", ".rna"), sort = F)
# first check the annotation match between distance and RNA meta
all(data_distance_erosion_raw_cell_rna$major.dna == data_distance_erosion_raw_cell_rna$major.rna)
all(data_distance_erosion_raw_cell_rna$sub.rna == data_distance_erosion_raw_cell_rna$sub.rna)
# plot the scatter between nCount_RNA (UMI) nFeature_RNA and long/short ratio
panel_global_erosion_ls_vs_nCount_rna <- data_distance_erosion_raw_cell_rna %>%
  filter(!sub.dna %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub.dna = factor(sub.dna, levels = global_erosion_score_rank_sub$sub)) %>%
  ungroup() %>%
  #group_by(disease_group.dna, sub.dna) %>%
  mutate(scaled_nCount_RNA = scale(log2(nCount_RNA))) %>%
  ggplot(aes(x = scaled_erosion_score_ls, y = scaled_nCount_RNA, color = disease_group.dna)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  #geom_point(size = 0.5, alpha = 0.6) +
  geom_density_2d(bins = 12) +
  geom_xsideboxplot(aes(y = disease_group.dna), outlier.shape = NA, orientation = "y") +
  geom_ysideboxplot(aes(x = disease_group.dna), outlier.shape = NA, orientation = "x") +
  #facet_grid(disease_group.dna ~ sub.dna) +
  facet_wrap(~sub.dna, ncol = 5) +
  xlab("zscore - log2 long/short interaction ratio") +
  ylab("zscore - log2 (nCount_RNA (UMI))") +
  scale_color_manual(name = "", values = c("AD" = "#f1a340", "CT" = "#998ec3")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 0, hjust = 0.5, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none") +
  theme(ggside.panel.scale = .3)
panel_global_erosion_ls_vs_scaled <- data_distance_erosion_raw_cell_rna %>%
  filter(!sub.dna %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub.dna = factor(sub.dna, levels = global_erosion_score_rank_sub$sub)) %>%
  ungroup() %>%
  #group_by(disease_group.dna, sub.dna) %>%
  mutate(scaled_scale = scale(scaled)) %>%
  ggplot(aes(x = scaled_erosion_score_ls, y = scaled_scale, color = disease_group.dna)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  #geom_point(size = 0.5, alpha = 0.6) +
  geom_density_2d(bins = 12) +
  geom_xsideboxplot(aes(y = disease_group.dna), outlier.shape = NA, orientation = "y") +
  geom_ysideboxplot(aes(x = disease_group.dna), outlier.shape = NA, orientation = "x") +
  #facet_grid(disease_group.dna ~ sub.dna) +
  facet_wrap(~sub.dna, ncol = 5) +
  xlab("zscore - log2 long/short interaction ratio") +
  ylab("zscore - scaled") +
  scale_color_manual(name = "", values = c("AD" = "#f1a340", "CT" = "#998ec3")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 0, hjust = 0.5, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none") +
  theme(ggside.panel.scale = .3)
panel_global_erosion_ls_vs_nFeature_rna <- data_distance_erosion_raw_cell_rna %>%
  filter(!sub.dna %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub.dna = factor(sub.dna, levels = global_erosion_score_rank_sub$sub)) %>%
  ungroup() %>%
  #group_by(disease_group.dna, sub.dna) %>%
  mutate(scaled_nFeature_RNA = scale(log2(nFeature_RNA))) %>%
  ggplot(aes(x = scaled_erosion_score_ls, y = scaled_nFeature_RNA, color = disease_group.dna)) +
  geom_hline(yintercept = 0.0, linetype = 2, color = "black") +
  geom_vline(xintercept = 0.0, linetype = 2, color = "black") +
  #geom_point(size = 0.5, alpha = 0.6) +
  geom_density_2d(bins = 12) +
  geom_xsideboxplot(aes(y = disease_group.dna), outlier.shape = NA, orientation = "y") +
  geom_ysideboxplot(aes(x = disease_group.dna), outlier.shape = NA, orientation = "x") +
  #facet_grid(disease_group.dna ~ sub.dna) +
  facet_wrap(~sub.dna, ncol = 5) +
  xlab("zscore - log2 long/short interaction ratio") +
  ylab("zscore - log2 (nFeature_RNA (Gene))") +
  scale_color_manual(name = "", values = c("AD" = "#f1a340", "CT" = "#998ec3")) +
  theme_cowplot() +
  theme(strip.clip = 'off', strip.background = element_blank()) +
  theme(strip.text.x = element_text(face = "bold", angle = 0, hjust = 0.5, vjust = 0)) +
  theme(strip.text.y = element_text(face = "bold", angle = 0)) +
  theme(legend.position = "none") +
  theme(ggside.panel.scale = .3)
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_scatter_RNA.pdf", height = 10, width = 12)
print(panel_global_erosion_ls_vs_nCount_rna)
print(panel_global_erosion_ls_vs_nFeature_rna)
print(panel_global_erosion_ls_vs_scaled)
dev.off()
########################
# exploration II-D: distance vs contact frequency cluster (smoothed)
########################
# --- functions
get_sorted_distance_distr.cluster <- function(data_distance_percent_smooth, data_distance_erosion_cell, cluster_order_by_erosion) {
  # construct cell by bin data.frame
  tmp <- data_distance_percent_smooth
  tmp$cluster_id <- data_distance_erosion_cell$cluster_id[match(tmp$cell_id, data_distance_erosion_cell$cell_id)]
  # build the percentage matrix
  tmp_matrix <- tmp %>% 
    ungroup() %>%
    filter(!is.na(cluster_id)) %>%
    select(cluster_id, bin_idx, count, total_count) %>%
    group_by(cluster_id, bin_idx) %>%
    summarise(count = sum(count), total_count = sum(total_count)) %>%
    mutate(percent = count / total_count * 100) %>%
    select(cluster_id, bin_idx, percent) %>%
    pivot_wider(names_from = "bin_idx", values_from = "percent")
  # sort the rows by cluster order
  tmp_matrix_sorted <- tmp_matrix %>%
    mutate(cluster_id = factor(cluster_id, levels = cluster_order_by_erosion)) %>%
    arrange(cluster_id)
  # confirm the order of cluster id is the same as given
  cat(paste0("The order of cell-cluster match with given order: ", all(tmp_matrix_sorted$cluster_id == cluster_order_by_erosion), "\n"))
  # get percentage matrix
  tmp_mat <- as.matrix(tmp_matrix_sorted[, 2:ncol(tmp_matrix_sorted)])
  #tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  row.names(tmp_mat) <- tmp_matrix_sorted$cluster_id
  # reverse the order of columns
  tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  #tmp_mat <- t(tmp_mat)
  return(tmp_mat)
}
# --- end
# filter cell types then sort the cell into decile by scaled_erosion_score 
data_distance_erosion_smooth_cell_ntile <- data_distance_erosion_smooth_cell %>%
  filter(major %in% c("Astro", "Endo", "Exc", "Inh", "Micro", "Oligo", "OPC", "VLMC")) %>%
  #group_by(erosion_label, sub) %>%
  mutate(erosion_label = erosion_label_ls, erosion_score = erosion_score_ls) %>%
  select(cell_id, erosion_label, erosion_score, gender, major, sub, disease_group) %>%
  group_by(erosion_label, sub, disease_group) %>%
  mutate(erosion_group =  ntile(erosion_score, 20)) %>%
  #group_by(erosion_label, sub, erosion_group) %>%
  group_by(erosion_label, sub, disease_group, erosion_group) %>%
  mutate(total_count = n()) %>%
  #female_count = sum(gender == "female"), 
  #disease_count = sum(disease_group == "AD"))  %>%
  #mutate(cluster_id = paste0(sub, '.C_', erosion_group))
  mutate(cluster_id = paste0(sub, '.', disease_group, '.C_', erosion_group))
# cluster
data_distance_erosion_smooth_cluster <- data_distance_erosion_smooth_cell_ntile %>%
  group_by(erosion_label, sub, disease_group, erosion_group, cluster_id, total_count) %>%
  summarise(cluster_erosion_score = mean(erosion_score)) %>%
  #cluster_pmi = mean(pmi),
  #cluster_female_frac = mean(female_count / total_count),
  #cluster_disease_frac = mean(disease_count / total_count)) %>%
  #group_by(erosion_label, sub) %>%
  group_by(erosion_label, sub, disease_group) %>%
  arrange(cluster_erosion_score, .by_group = T) 
cluster_order_by_erosion <- data_distance_erosion_smooth_cluster$cluster_id
# build the distance percentage table
data_distance_percent <- get_sorted_distance_distr.cluster(data_distance_percent_smooth, data_distance_erosion_smooth_cell_ntile, cluster_order_by_erosion)
# sort cell meta information by the same order
data_cluster_hic_erosion_sort <- data_distance_erosion_smooth_cluster %>%
  group_by() %>%
  mutate(cluster_id = factor(cluster_id, levels = cluster_order_by_erosion)) %>%
  mutate(cell_group = ifelse(sub %in% c("Astro", "Endo", "Micro", "OPC", "Oligo", "VLMC"), "Glial", "Neurons")) %>%
  arrange(cluster_id, .by_group = T)

########################
# plot zoom-in view and aggregated plot - smoothed
########################
# construct cell by bin data.frame
data_distance_percent_smooth_zoom <- data_distance_percent_smooth
data_distance_percent_smooth_zoom$cluster_id <- data_distance_erosion_smooth_cell_ntile$cluster_id[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
data_distance_percent_smooth_zoom$erosion_group <- data_distance_erosion_smooth_cell_ntile$erosion_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
data_distance_percent_smooth_zoom$sub <- data_distance_erosion_smooth_cell_ntile$sub[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
data_distance_percent_smooth_zoom$major <- data_distance_erosion_smooth_cell_ntile$major[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
data_distance_percent_smooth_zoom$disease_group <- data_distance_erosion_smooth_cell_ntile$disease_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
data_distance_percent_smooth_zoom$gender <- data_distance_erosion_smooth_cell_ntile$gender[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
tmp_bin_range <- unique(data_distance_percent_smooth_zoom[, c("bin_idx", "bin_start")])
tmp_bin_range_filter <- subset(tmp_bin_range, bin_idx %in% c(0, 4, 8, 12, 16, 20, 24, 28, 32))
tmp_bin_range_filter_label <- c("1 kb", "4 kb", "16 kb", "64 kb", "256 kb", "1.024 Mb", "4.096 Mb", "16.38 Mb", "65.54 Mb")
# build the percentage matrix zoom-in view - sub
panel_distance_percent_zoom_in_tmp <- data_distance_percent_smooth_zoom %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_diff <- panel_distance_percent_zoom_in_tmp %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  #theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in <- panel_distance_percent_zoom_in_tmp %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  xlab("Erosion score group") +
  ylab("Distance group") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#0571b0", "#f7f7f7", "#ca0020"), values = scales::rescale(c(0, 4, 6)), guide = "colorbar", limits=c(0, 6), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_blank()) +
  #theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
# build the percentage matrix aggregating plot - major
panel_distance_percent_summary <- data_distance_percent_smooth_zoom %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(major, disease_group, bin_idx, count, total_count) %>%
  group_by(major, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count) %>%
  ggplot(aes(x = bin_idx, y = percent, color = disease_group, fill = disease_group)) +
  #geom_line(linewidth = 1, alpha = 0.7) +
  geom_area(alpha = 0.9, position = 'identity') +
  geom_hline(yintercept = 0.0, color = "black") +
  xlab("Distance group") +
  ylab("Percent") +
  facet_wrap(~ major, nrow = 2) +
  scale_color_manual(name = "", values = c("AD" = "#d01c8b", "CT" = "#4dac26")) +
  scale_fill_manual(name = "", values = c("AD" = "#f1b6da", "CT" = "#b8e186")) +
  scale_x_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx) +
  scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  theme_cowplot() +
  theme(strip.background = element_blank())
# build the percentage matrix aggregating plot - sub
panel_distance_percent_summary_sub <- data_distance_percent_smooth_zoom %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(sub, disease_group, bin_idx, count, total_count) %>%
  group_by(sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count) %>%
  ggplot(aes(x = bin_idx, y = percent, color = disease_group, fill = disease_group)) +
  #geom_line(linewidth = 1, alpha = 0.7) +
  geom_area(alpha = 0.9, position = 'identity') +
  geom_hline(yintercept = 0.0, color = "black") +
  xlab("Distance group") +
  ylab("Percent") +
  facet_wrap(~ sub, nrow = 2) +
  scale_color_manual(name = "", values = c("AD" = "#d01c8b", "CT" = "#4dac26")) +
  scale_fill_manual(name = "", values = c("AD" = "#f1b6da", "CT" = "#b8e186")) +
  scale_x_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx) +
  scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  theme_cowplot() +
  theme(strip.background = element_blank())
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_zoom_in.pdf", width = 8, height = 50)
plot_grid(panel_distance_percent_zoom_in, panel_distance_percent_zoom_in_diff, align = "h", nrow =1, axis = "lr", rel_widths = c(2,2))
dev.off()
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_aggregation.pdf", width = 6, height = 4)
print(panel_distance_percent_summary)
dev.off()
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_aggregation.sub.pdf", width = 16, height = 4)
print(panel_distance_percent_summary_sub)
dev.off()
########################
# plot the Circos plot
########################
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_circos.pdf", height = 12, width = 12)
circos.clear()
# par
circos.par("start.degree" = 68.8, track.margin = c(0.005, 0.005))
# initialize
#circos.heatmap.initialize(data_distance_percent, split = data_cluster_hic_erosion_sort$sub)
col_neuron = c("Glial" = "#e66101", "Neurons" = "#5e3c99")
circos.heatmap(data_cluster_hic_erosion_sort$cell_group, split = data_cluster_hic_erosion_sort$sub, cluster = F, col = col_neuron, track.height = 0.02, show.sector.labels = TRUE, )
# distance distribution
col_fun_heatmap = colorRamp2(c(0, 4, 6), c("#0571b0", "#f7f7f7", "#ca0020"))
circos.heatmap(data_distance_percent,  split = data_cluster_hic_erosion_sort$sub, cluster = F, col = col_fun_heatmap, show.sector.labels = FALSE, track.height = 0.12)
# AD and CT
col_disease_group = c("AD" = "#d01c8b", "CT" = "#f1b6da")
circos.heatmap(data_cluster_hic_erosion_sort$disease_group, col = col_disease_group, track.height = 0.02)
# disease_frac
#col_disease_frac = colorRamp2(c(0.3, 0.7), c("#f7f7f7", "#fb8072"))
#circos.heatmap(data_cluster_hic_erosion_sort$cluster_disease_frac, col = col_disease_frac, track.height = 0.03)
# erosion score
col_erosion = colorRamp2(c(-1.5, 0, 2.5), c("#008837", "#f7f7f7", "#7b3294"))
circos.heatmap(data_cluster_hic_erosion_sort$cluster_erosion_score, col = col_erosion, track.height = 0.02)
# female_frac
col_female_frac = colorRamp2(c(0.3, 0.7), c("#f7f7f7", "#404040"))
#circos.heatmap(data_cluster_hic_erosion_sort$cluster_female_frac, col = col_female_frac, track.height = 0.02)
# add legend
lgd_heatmap = Legend(title = "Percent of interaction", col_fun = col_fun_heatmap)
lgd_disease_group = Legend(title = "Disease group", at = names(col_disease_group), legend_gp = gpar(fill = col_disease_group))
lgd_neuron = Legend(title = "Cell group", at = names(col_neuron), legend_gp = gpar(fill = col_neuron))
lgd_erosion = Legend(title = "erosion score", col_fun = col_erosion)
lgd_female_frac = Legend(title = "Percent of female", col_fun = col_female_frac)
lgd_list = packLegend(lgd_heatmap, lgd_disease_group, lgd_neuron)
grid.draw(lgd_list)
dev.off()
# disease_frac
#col_disease_frac = colorRamp2(c(0.3, 0.5, 0.7), c("#0571b0", "#f7f7f7", "#ca0020"))
#circos.track(ylim = c(-2, 2), panel.fun = function(x, y) {
#    y = data_cluster_hic_erosion_sort$cluster_disease_ratio[CELL_META$subset]
#    y = y[CELL_META$row_order]
#    circos.lines(seq_along(y) - 0.5, y, type = 'h')
#})
```


############################
# long / short distribution - by chrom -- manuscript
############################
```{r}
library(ggplot2)
library(ggside)
library(dplyr)
library(tidyr)
library(cowplot)
library(ggsignif)
#######################
# function 
#######################
# load cell annotation
load_cell_annotation <- function(filename) {
  anno_cell <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
  # add major cell type,
  anno_cell <- anno_cell %>%
    mutate(sub2 = sub) %>%
    separate(sub2, sep = ' ', into = c("major", "Extra"), extra = "merge") %>% 
    select(-Extra)
  return(anno_cell)
}
# load sample annotation
load_sample_annotation <- function(filename) {
  anno_sample <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
}
# load distance track 
load_read_pair_distance_summary_by_chrom <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".by_chrom.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)

    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
#######################
# load data & annotation
#######################
# cell annotation
#anno_cell <- load_cell_annotation("results/202501/cell_meta_table.RNA.tsv")
anno_cell <- read.table(file = "results/202501/cell_meta_table.tsv", sep = '\t', header = T, stringsAsFactors = F)
# sample annotation
anno_sample <- load_sample_annotation("annotation/sample_info.v2.tsv")
# single-cell distance summary
patient_list_20 <- NULL 
# due to privary issue, sample_ids are not provided here
data_pair_distance_mattew_by_chrom <- load_read_pair_distance_summary_by_chrom("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
#######################
# analysis and plot
#######################
data_distance_erosion_by_chrom <- data_pair_distance_mattew_by_chrom %>%
  filter(bin_idx != "inter") %>%
  group_by(cell_id, chrom) %>%
  mutate(percent = count/sum(count)) %>%
  mutate(distance_group = ifelse(bin_start <= 128000, paste0("range_", round(bin_start/1000, 3), "kb", "_", round(bin_end/1000, 3), "kb"), paste0("range_", round(bin_start/1000000, 3), "Mb", "_", round(bin_end/1000000, 3), "Mb"))) %>%
  select(cell_id, chrom, distance_group, percent) %>%
  pivot_wider(names_from = "distance_group", values_from = "percent") 
# merge with cell 
data_distance_erosion_by_chrom <- merge(data_distance_erosion_by_chrom, anno_cell[, c("cell_id", "major", "sub", "disease_group", "gender", "projid")], by = "cell_id", sort = F)
# calculate erosion score
data_distance_erosion_by_chrom_zscore <- data_distance_erosion_by_chrom %>%
  filter(range_1kb_128kb > 0 & range_128kb_4096kb > 0 & range_4.096Mb_65.536Mb > 0) %>%
  mutate(erosion_score_ls = log2(range_4.096Mb_65.536Mb/range_1kb_128kb),
         erosion_score_lm = log2(range_4.096Mb_65.536Mb/range_128kb_4096kb),
         erosion_score_ms = log2(range_128kb_4096kb/range_1kb_128kb)) 
data_distance_erosion_by_chrom_zscore$erosion_score_ls_zscore <- scale(data_distance_erosion_by_chrom_zscore$erosion_score_ls, center = T, scale = T)
data_distance_erosion_by_chrom_zscore$erosion_score_lm_zscore <- scale(data_distance_erosion_by_chrom_zscore$erosion_score_lm, center = T, scale = T)
data_distance_erosion_by_chrom_zscore$erosion_score_ms_zscore <- scale(data_distance_erosion_by_chrom_zscore$erosion_score_ms, center = T, scale = T)
# 
global_erosion_score_rank_major <- data_distance_erosion_by_chrom_zscore %>%
  group_by(major) %>%
  filter(chrom == "chr1") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
global_erosion_score_rank_sub <- data_distance_erosion_by_chrom_zscore %>%
  group_by(sub) %>%
  filter(chrom == "chr1") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
#
panel_global_erosion_ls_boxplot_cell_chrom_example <- data_distance_erosion_by_chrom_zscore %>%
    mutate(chrom = factor(chrom, levels = c(paste0("chr", seq(1,22)), "chrX"))) %>%
    filter(chrom %in% c('chr1', 'chr8')) %>%
    mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
    ggplot(aes(x = sub, y = erosion_score_ls_zscore, fill = sub, alpha = disease_group)) +
    geom_boxplot(outlier.shape = NA, width = 0.6, position = position_dodge(width = 0.7)) +
    xlab("") +
    ylab("log2 long vs short interaction") +
    scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                            "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                            "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                            "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                            "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                            "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                            "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
    facet_wrap(~chrom, ncol = 1) +
    theme_cowplot() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    theme(legend.position = "none") +
    theme(strip.background = element_blank())
#
panel_global_erosion_ls_summary_cell_chrom <- data_distance_erosion_by_chrom_zscore %>%
  mutate(chrom = factor(chrom, levels = c(paste0("chr", seq(1,22)), "chrX"))) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(chrom, sub, disease_group) %>%
  summarise(mean_erosion_score_ls_zscore = mean(erosion_score_ls_zscore, na.rm = T)) %>%
  select(chrom, sub, disease_group, mean_erosion_score_ls_zscore) %>%
  pivot_wider(names_from = "disease_group", values_from = "mean_erosion_score_ls_zscore") %>%
  ggplot(aes(x = chrom, y = sub, fill = AD-CT)) +
  geom_tile(color = "black") +
  xlab("") +
  ylab("") +
  scale_fill_gradientn(colours = c("#0571b0", "#f7f7f7", "#ca0020"), values = scales::rescale(c(-0.5, 0, 0.5)), guide = "colorbar", limits=c(-0.5, 0.5), oob = scales::squish) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "right")
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.by_chrom.pdf", width = 11, height = 7)
print(panel_global_erosion_ls_boxplot_cell_chrom_example)
print(panel_global_erosion_ls_summary_cell_chrom)
dev.off()

```


############################
# long / short distribution - CTCF -- manuscript
############################
```{r}
library(ggplot2)
library(ggside)
library(dplyr)
library(tidyr)
library(cowplot)
#######################
# function 
#######################
# load cell annotation
load_cell_annotation <- function(filename) {
  anno_cell <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
  # # add major cell type,
  # anno_cell <- anno_cell %>%
  #   mutate(sub2 = sub) %>%
  #   separate(sub2, sep = ' ', into = c("major", "Extra"), extra = "merge") %>% 
  #   select(-Extra)
  return(anno_cell)
}
# load sample annotation
load_sample_annotation <- function(filename) {
  anno_sample <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
}
# load distance track 
load_read_pair_distance_summary_CTCF <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".pair_distance.ctcf.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)
    # fix the typo in colnames
    names(table)[names(table) == "sampole_id"] <- "sample_id"
    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
load_read_pair_distance_summary_CTCF_loose <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".loose.pair_distance.ctcf.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)
    # fix the typo in colnames
    names(table)[names(table) == "sampole_id"] <- "sample_id"
    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
#######################
# load data & annotation
#######################
# cell annotation
anno_cell <- load_cell_annotation("results/202501/cell_meta_table.RNA.tsv")
# sample annotation
anno_sample <- load_sample_annotation("annotation/sample_info.v2.tsv")
# single-cell distance summary

patient_list_20 <- NULL
# due to privary issue, sample_ids are not provided here 
data_pair_distance_mattew_strict <- load_read_pair_distance_summary_CTCF("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
data_pair_distance_mattew_loose <- load_read_pair_distance_summary_CTCF_loose("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
#######################
# exploration I-A: erosion score, short-mid-long range
#######################
# --- functions -- 
get_distance_smooth_CTCF <- function(data_pair_distance, agg_ratio = 4) {
  # construct cell by bin data.frame
  tmp <- data_pair_distance %>%
    # remove inter-chromosomal interaction
    filter(bin_idx != "inter") %>% 
    mutate(bin_idx = as.numeric(bin_idx)) %>%
    mutate(bin_idx = bin_idx %/% agg_ratio) %>%
    #mutate(bin_group = ifelse(ATAC_group %in% c("Peak-Peak", "Peak-Other"), "Peak", "Other")) %>%
    mutate(bin_group = ctcf_group) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx, bin_group) %>%
    mutate(bin_start = min(bin_start), bin_end = max(bin_end)) %>%
    select(cell_id, sample_id, distance_group, bin_idx, bin_group, bin_start, bin_end, count) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx, bin_group, bin_start, bin_end) %>%
    summarise(count = sum(count)) %>%
    group_by(cell_id, sample_id, distance_group, bin_group) %>%
    mutate(total_count = sum(count)) %>%
    mutate(percent = count / total_count * 100)
  return(tmp)
}
# --- end
# build the distance percentage table - binning by four bins
data_distance_percent_smooth_strict <- get_distance_smooth_CTCF(data_pair_distance_mattew_strict, agg_ratio = 4)
data_distance_percent_smooth_loose <- get_distance_smooth_CTCF(data_pair_distance_mattew_loose, agg_ratio = 4)
#######################
# exploration I-A: erosion score, short-mid-long range
#######################
# --- functions -- 
get_distance_matrix_smooth_erosion_CTCF <- function(data_distance_percent_smooth, sr_start, sr_end, lr_start, lr_end, agg_ratio = 4) {
  sr_start = sr_start * 4 / agg_ratio
  sr_end = sr_end * 4 / agg_ratio 
  lr_start = lr_start * 4 / agg_ratio
  lr_end = lr_end * 4 / agg_ratio 
  mid_start = sr_end + 1
  mid_end = lr_start - 1
  # get the bin_idx to bin range 
  bin_range <- unique(data_distance_percent_smooth[, c("bin_idx", "bin_start", "bin_end")])
  erosion_score_label_ls <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  erosion_score_label_lm <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", '(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')')
  erosion_score_label_ms <- paste0('(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')', " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  # get the erosion x cell id
  data_erosion <- data_distance_percent_smooth %>%
    mutate(range_group = ifelse(bin_idx >= sr_start - 1 & bin_idx <= sr_end - 1, 'short_range', 
                                ifelse(bin_idx >= mid_start - 1 & bin_idx <= mid_end - 1, 'mid_range',
                                       ifelse(bin_idx >= lr_start - 1 & bin_idx <= lr_end -1, 'long_range', 'other')))) %>%
    group_by(cell_id, bin_group, range_group, total_count) %>%
    summarise(percent =  sum(percent)) %>%
    pivot_wider(names_from = "range_group", values_from = "percent") %>%
    # add 1 pseudo count
    mutate(erosion_score_ls = log2((long_range + 0.1)/(short_range + 0.1)), 
           erosion_score_lm = log2((long_range + 0.1)/(mid_range + 0.1)),
           erosion_score_ms = log2((mid_range + 0.1)/(short_range + 0.1))) %>%
    group_by(bin_group) %>%
    mutate(scaled_erosion_score_ls = scale(erosion_score_ls),
           scaled_erosion_score_lm = scale(erosion_score_lm),
           scaled_erosion_score_ms = scale(erosion_score_ms))
  # add erosion label
  data_erosion$erosion_label_ls <- erosion_score_label_ls
  data_erosion$erosion_label_lm <- erosion_score_label_lm
  data_erosion$erosion_label_ms <- erosion_score_label_ms
  return(data_erosion)
}
# --- end
data_distance_erosion_smooth_strict <- get_distance_matrix_smooth_erosion_CTCF(data_distance_percent_smooth_strict, 1, 14, 25, 32, agg_ratio = 4)
data_distance_erosion_smooth_loose <- get_distance_matrix_smooth_erosion_CTCF(data_distance_percent_smooth_loose, 1, 14, 25, 32, agg_ratio = 4)
# merge with cell annotation
data_distance_erosion_smooth_cell_strict <- merge(data_distance_erosion_smooth_strict, anno_cell, by = "cell_id")
data_distance_erosion_smooth_cell_loose <- merge(data_distance_erosion_smooth_loose, anno_cell, by = "cell_id")
# save the results to files
write.table(data_distance_erosion_smooth_cell_strict, file = "results/202501/cool_file/distance_erosion_smooth.CTCF.strict.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
write.table(data_distance_erosion_smooth_cell_loose, file = "results/202501/cool_file/distance_erosion_smooth.CTCF.loose.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
data_distance_erosion_smooth_cell_strict <- read.table(file = "results/202501/cool_file/distance_erosion_smooth.CTCF.strict.tsv", sep = '\t', header = T)
data_distance_erosion_smooth_cell_loose <- read.table(file = "results/202501/cool_file/distance_erosion_smooth.CTCF.loose.tsv", sep = '\t', header = T)
########################
# exploration II-A: distance vs contact frequency cluster boxplot (sub-cell type)
########################
# step 1: get the overall long vs short range regardless of disease_group
global_erosion_score_rank_major <- data_distance_erosion_smooth_cell_strict %>%
  group_by(major) %>%
  filter(bin_group == "Other-Other") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
global_erosion_score_rank_sub <- data_distance_erosion_smooth_cell_strict %>%
  group_by(sub) %>%
  filter(bin_group == "Other-Other") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
# step 2: for AD and CT rank the cells by global rank major then by global rank sub
global_raw_erosion_ad <- data_distance_erosion_smooth_cell_strict %>%
  filter(disease_group == 'AD') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(bin_group, major, sub, erosion_score_ls)
global_raw_erosion_ct <- data_distance_erosion_smooth_cell_strict %>%
  filter(disease_group == 'CT') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(bin_group, major, sub, erosion_score_ls)
# group by disease_group, then by sub-cell type
panel_global_erosion_long_range_boxplot_cell_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = long_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_mid_range_boxplot_cell_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = mid_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of middle-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_short_range_boxplot_cell_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = short_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_long_range_boxplot_patient_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(long_range_ave = mean(long_range)) %>%
  ggplot(aes(x = disease_group, y = long_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_middle_range_boxplot_patient_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(mid_range_ave = mean(mid_range)) %>%
  ggplot(aes(x = disease_group, y = mid_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of mid-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_short_range_boxplot_patient_strict <- data_distance_erosion_smooth_cell_strict %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(short_range_ave = mean(short_range)) %>%
  ggplot(aes(x = disease_group, y = short_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("CTCF (stringent threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
#
panel_global_erosion_long_range_boxplot_cell_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = long_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_mid_range_boxplot_cell_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = mid_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of middle-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_short_range_boxplot_cell_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = c("Other-Other", "CTCF-Other", "CTCF-CTCF"))) %>%
  ggplot(aes(x = bin_group, y = short_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_long_range_boxplot_patient_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(long_range_ave = mean(long_range)) %>%
  ggplot(aes(x = disease_group, y = long_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_middle_range_boxplot_patient_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(mid_range_ave = mean(mid_range)) %>%
  ggplot(aes(x = disease_group, y = mid_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of mid-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_short_range_boxplot_patient_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(short_range_ave = mean(short_range)) %>%
  ggplot(aes(x = disease_group, y = short_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("CTCF (loose threshold)") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
#
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.CTCF.pdf", height = 12, width = 18)
print(panel_global_erosion_long_range_boxplot_cell_strict)
print(panel_global_erosion_mid_range_boxplot_cell_strict)
print(panel_global_erosion_short_range_boxplot_cell_strict)
print(panel_global_erosion_long_range_boxplot_patient_strict)
print(panel_global_erosion_middle_range_boxplot_patient_strict)
print(panel_global_erosion_short_range_boxplot_patient_strict)
print(panel_global_erosion_long_range_boxplot_cell_loose)
print(panel_global_erosion_mid_range_boxplot_cell_loose)
print(panel_global_erosion_short_range_boxplot_cell_loose)
print(panel_global_erosion_long_range_boxplot_patient_loose)
print(panel_global_erosion_middle_range_boxplot_patient_loose)
print(panel_global_erosion_short_range_boxplot_patient_loose)
dev.off()
#
panel_global_erosion_short_range_heatmap_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = rev(c("Other-Other", "CTCF-Other", "CTCF-CTCF")))) %>%
  group_by(bin_group, sub, disease_group) %>%
  summarise(short_range_mean = mean(short_range, na.rm = T)) %>%
  pivot_wider(names_from = "disease_group", values_from = "short_range_mean") %>%
  ggplot(aes(x = sub, y = bin_group, fill = (AD-CT)/100)) +
  geom_tile(color = "black", linewidth = 0.5) +
  xlab("") +
  ylab("") +
  ggtitle("Percent of short-range interaction - CTCF loose") +
  scale_fill_gradientn(name = "AD - non-AD", colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-0.03, 0, 0.03)), guide = "colorbar", limits=c(-0.03, 0.03), oob = scales::squish , labels = scales::percent) +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.line = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.border = element_rect(color = "black", linewidth = 1))
panel_global_erosion_mid_range_heatmap_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = rev(c("Other-Other", "CTCF-Other", "CTCF-CTCF")))) %>%
  group_by(bin_group, sub, disease_group) %>%
  summarise(mid_range_mean = mean(mid_range, na.rm = T)) %>%
  pivot_wider(names_from = "disease_group", values_from = "mid_range_mean") %>%
  ggplot(aes(x = sub, y = bin_group, fill = (AD-CT)/100)) +
  geom_tile(color = "black", linewidth = 0.5) +
  xlab("") +
  ylab("") +
  ggtitle("Percent of mid-range interaction - CTCF loose") +
  scale_fill_gradientn(name = "AD - non-AD", colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-0.02, 0, 0.02)), guide = "colorbar", limits=c(-0.02, 0.02), oob = scales::squish , labels = scales::percent) +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.line = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.border = element_rect(color = "black", linewidth = 1))
panel_global_erosion_long_range_heatmap_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  mutate(bin_group = factor(bin_group, levels = rev(c("Other-Other", "CTCF-Other", "CTCF-CTCF")))) %>%
  group_by(bin_group, sub, disease_group) %>%
  summarise(long_range_mean = mean(long_range, na.rm = T)) %>%
  pivot_wider(names_from = "disease_group", values_from = "long_range_mean") %>%
  ggplot(aes(x = sub, y = bin_group, fill = (AD-CT)/100)) +
  geom_tile(color = "black", linewidth = 0.5) +
  xlab("") +
  ylab("") +
  ggtitle("Percent of long-range interaction - CTCF loose") +
  scale_fill_gradientn(name = "AD - non-AD", colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-0.03, 0, 0.03)), guide = "colorbar", limits=c(-0.03, 0.03), oob = scales::squish , labels = scales::percent) +
  scale_y_discrete(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_cowplot() +
  theme(axis.line = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.border = element_rect(color = "black", linewidth = 1))

pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_heatmap.CTCF.pdf", height = 8, width = 15)
print(panel_global_erosion_short_range_heatmap_loose)
print(panel_global_erosion_mid_range_heatmap_loose)
print(panel_global_erosion_long_range_heatmap_loose)
dev.off()

########################
# exploration II-D: distance vs contact frequency cluster (smoothed) - sub
########################
# --- functions
get_sorted_distance_distr.cluster <- function(data_distance_percent_smooth, data_distance_erosion_cell, data_distance_erosion_smooth_cluster, bin_group_label) {
  cluster_order_by_erosion <- subset(data_distance_erosion_smooth_cluster, bin_group == bin_group_label)$cluster_id
  # construct cell by bin data.frame
  tmp <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  tmp$cluster_id <- data_distance_erosion_cell$cluster_id[match(tmp$cell_id, data_distance_erosion_cell$cell_id)]
  # build the percentage matrix
  tmp_matrix <- tmp %>% 
    ungroup() %>%
    filter(!is.na(cluster_id)) %>%
    select(cluster_id, bin_idx, count, total_count) %>%
    group_by(cluster_id, bin_idx) %>%
    summarise(count = sum(count), total_count = sum(total_count)) %>%
    mutate(percent = count / total_count * 100) %>%
    select(cluster_id, bin_idx, percent) %>%
    pivot_wider(names_from = "bin_idx", values_from = "percent")
  # sort the rows by cluster order
  tmp_matrix_sorted <- tmp_matrix %>%
    mutate(cluster_id = factor(cluster_id, levels = cluster_order_by_erosion)) %>%
    arrange(cluster_id)
  # confirm the order of cluster id is the same as given
  cat(paste0("The order of cell-cluster match with given order: ", all(tmp_matrix_sorted$cluster_id == cluster_order_by_erosion), "\n"))
  # get percentage matrix
  tmp_mat <- as.matrix(tmp_matrix_sorted[, 2:ncol(tmp_matrix_sorted)])
  #tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  row.names(tmp_mat) <- tmp_matrix_sorted$cluster_id
  # reverse the order of columns
  tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  #tmp_mat <- t(tmp_mat)
  return(tmp_mat)
}
# --- end
# filter cell types then sort the cell into decile by scaled_erosion_score 
data_distance_erosion_smooth_cell_ntile_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(major %in% c("Astro", "Endo", "Exc", "Inh", "Micro", "Oligo", "OPC", "VLMC")) %>%
  #group_by(erosion_label, sub) %>%
  mutate(erosion_label = erosion_label_ls, erosion_score = erosion_score_ls) %>%
  filter(!is.na(erosion_score)) %>%
  select(bin_group, cell_id, erosion_label, erosion_score, gender, major, sub, disease_group) %>%
  group_by(bin_group, erosion_label, sub, disease_group) %>%
  mutate(erosion_group =  ntile(erosion_score, 20)) %>%
  group_by(bin_group, erosion_label, sub, disease_group, erosion_group) %>%
  mutate(total_count = n()) %>%
  mutate(cluster_id = paste0(sub, '.', disease_group, '.C_', erosion_group))
# cluster
data_distance_erosion_smooth_cluster_loose <- data_distance_erosion_smooth_cell_ntile_loose %>%
  group_by(bin_group, erosion_label, sub, disease_group, erosion_group, cluster_id, total_count) %>%
  summarise(cluster_erosion_score = mean(erosion_score)) %>%
  group_by(bin_group, erosion_label, sub, disease_group) %>%
  arrange(cluster_erosion_score, .by_group = T) 
# build the distance percentage table
data_distance_percent_loose_other_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'Other-Other')
data_distance_percent_loose_peak_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'CTCF-Other')
data_distance_percent_loose_peak_vs_peak <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'CTCF-CTCF')
########################
# plot zoom-in view and aggregated plot - smoothed (sub)
########################
# --- functions
build_zoom_in_view <- function(data_distance_percent_smooth, data_distance_erosion_smooth_cell_ntile, bin_group_label) {
  data_distance_percent_smooth <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  data_distance_erosion_smooth_cell_ntile <- subset(data_distance_erosion_smooth_cell_ntile, bin_group == bin_group_label)
  #
  data_distance_percent_smooth_zoom <- data_distance_percent_smooth
  data_distance_percent_smooth_zoom$cluster_id <- data_distance_erosion_smooth_cell_ntile$cluster_id[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$erosion_group <- data_distance_erosion_smooth_cell_ntile$erosion_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$sub <- data_distance_erosion_smooth_cell_ntile$sub[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$major <- data_distance_erosion_smooth_cell_ntile$major[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$disease_group <- data_distance_erosion_smooth_cell_ntile$disease_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$gender <- data_distance_erosion_smooth_cell_ntile$gender[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  return(data_distance_percent_smooth_zoom)
}
# --- end
data_distance_percent_smooth_zoom_loose_other_vs_other <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'Other-Other')
data_distance_percent_smooth_zoom_loose_peak_vs_other <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'CTCF-Other')
data_distance_percent_smooth_zoom_loose_peak_vs_peak <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'CTCF-CTCF')
#
tmp_bin_range <- unique(data_distance_percent_smooth_zoom_loose_other_vs_other[, c("bin_idx", "bin_start")])
tmp_bin_range_filter <- subset(tmp_bin_range, bin_idx %in% c(0, 4, 8, 12, 16, 20, 24, 28, 32))
tmp_bin_range_filter_label <- c("1 kb", "4 kb", "16 kb", "64 kb", "256 kb", "1.024 Mb", "4.096 Mb", "16.38 Mb", "65.54 Mb")
# build the percentage matrix zoom-in view - sub
panel_distance_percent_zoom_in_tmp_loose_other_vs_other <- data_distance_percent_smooth_zoom_loose_other_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_loose_peak_vs_other <- data_distance_percent_smooth_zoom_loose_peak_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_loose_peak_vs_peak <- data_distance_percent_smooth_zoom_loose_peak_vs_peak %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_diff_loose_other_vs_other <- panel_distance_percent_zoom_in_tmp_loose_other_vs_other %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Other vs Other") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_loose_peak_vs_other <- panel_distance_percent_zoom_in_tmp_loose_peak_vs_other %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Other") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_loose_peak_vs_peak <- panel_distance_percent_zoom_in_tmp_loose_peak_vs_peak %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Peak") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_zoom_in.CTCF.pdf", height = 50, width = 10)
plot_grid(panel_distance_percent_zoom_in_diff_loose_other_vs_other, panel_distance_percent_zoom_in_diff_loose_peak_vs_other, panel_distance_percent_zoom_in_diff_loose_peak_vs_peak, nrow = 1)
dev.off()


########################
# exploration II-E: distance vs contact frequency cluster (smoothed) - major
########################
# --- functions
get_sorted_distance_distr.cluster <- function(data_distance_percent_smooth, data_distance_erosion_cell, data_distance_erosion_smooth_cluster, bin_group_label) {
  cluster_order_by_erosion <- subset(data_distance_erosion_smooth_cluster, bin_group == bin_group_label)$cluster_id
  # construct cell by bin data.frame
  tmp <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  tmp$cluster_id <- data_distance_erosion_cell$cluster_id[match(tmp$cell_id, data_distance_erosion_cell$cell_id)]
  # build the percentage matrix
  tmp_matrix <- tmp %>% 
    ungroup() %>%
    filter(!is.na(cluster_id)) %>%
    select(cluster_id, bin_idx, count, total_count) %>%
    group_by(cluster_id, bin_idx) %>%
    summarise(count = sum(count), total_count = sum(total_count)) %>%
    mutate(percent = count / total_count * 100) %>%
    select(cluster_id, bin_idx, percent) %>%
    pivot_wider(names_from = "bin_idx", values_from = "percent")
  # sort the rows by cluster order
  tmp_matrix_sorted <- tmp_matrix %>%
    mutate(cluster_id = factor(cluster_id, levels = cluster_order_by_erosion)) %>%
    arrange(cluster_id)
  # confirm the order of cluster id is the same as given
  cat(paste0("The order of cell-cluster match with given order: ", all(tmp_matrix_sorted$cluster_id == cluster_order_by_erosion), "\n"))
  # get percentage matrix
  tmp_mat <- as.matrix(tmp_matrix_sorted[, 2:ncol(tmp_matrix_sorted)])
  #tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  row.names(tmp_mat) <- tmp_matrix_sorted$cluster_id
  # reverse the order of columns
  tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  #tmp_mat <- t(tmp_mat)
  return(tmp_mat)
}
# --- end
# filter cell types then sort the cell into decile by scaled_erosion_score 
data_distance_erosion_smooth_cell_ntile_loose <- data_distance_erosion_smooth_cell_loose %>%
  filter(major %in% c("Astro", "Endo", "Exc", "Inh", "Micro", "Oligo", "OPC", "VLMC")) %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  #group_by(erosion_label, sub) %>%
  mutate(erosion_label = erosion_label_ls, erosion_score = erosion_score_ls) %>%
  filter(!is.na(erosion_score)) %>%
  select(bin_group, cell_id, erosion_label, erosion_score, gender, major, major, disease_group) %>%
  group_by(bin_group, erosion_label, major, disease_group) %>%
  mutate(erosion_group =  ntile(erosion_score, 20)) %>%
  group_by(bin_group, erosion_label, major, disease_group, erosion_group) %>%
  mutate(total_count = n()) %>%
  mutate(cluster_id = paste0(major, '.', disease_group, '.C_', erosion_group))
# cluster
data_distance_erosion_smooth_cluster_loose <- data_distance_erosion_smooth_cell_ntile_loose %>%
  group_by(bin_group, erosion_label, major, disease_group, erosion_group, cluster_id, total_count) %>%
  summarise(cluster_erosion_score = mean(erosion_score)) %>%
  group_by(bin_group, erosion_label, major, disease_group) %>%
  arrange(cluster_erosion_score, .by_group = T) 
# build the distance percentage table
data_distance_percent_loose_other_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'Other-Other')
data_distance_percent_loose_peak_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'CTCF-Other')
data_distance_percent_loose_peak_vs_peak <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, data_distance_erosion_smooth_cluster_loose, 'CTCF-CTCF')
########################
# plot zoom-in view and aggregated plot - smoothed (sub)
########################
# --- functions
build_zoom_in_view <- function(data_distance_percent_smooth, data_distance_erosion_smooth_cell_ntile, bin_group_label) {
  data_distance_percent_smooth <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  data_distance_erosion_smooth_cell_ntile <- subset(data_distance_erosion_smooth_cell_ntile, bin_group == bin_group_label)
  #
  data_distance_percent_smooth_zoom <- data_distance_percent_smooth
  data_distance_percent_smooth_zoom$cluster_id <- data_distance_erosion_smooth_cell_ntile$cluster_id[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$erosion_group <- data_distance_erosion_smooth_cell_ntile$erosion_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  #data_distance_percent_smooth_zoom$major <- data_distance_erosion_smooth_cell_ntile$major[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$major <- data_distance_erosion_smooth_cell_ntile$major[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$disease_group <- data_distance_erosion_smooth_cell_ntile$disease_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$gender <- data_distance_erosion_smooth_cell_ntile$gender[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  return(data_distance_percent_smooth_zoom)
}
# --- end
data_distance_percent_smooth_zoom_loose_other_vs_other <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'Other-Other')
data_distance_percent_smooth_zoom_loose_peak_vs_other <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'CTCF-Other')
data_distance_percent_smooth_zoom_loose_peak_vs_peak <- build_zoom_in_view(data_distance_percent_smooth_loose, data_distance_erosion_smooth_cell_ntile_loose, 'CTCF-CTCF')
#
tmp_bin_range <- unique(data_distance_percent_smooth_zoom_loose_other_vs_other[, c("bin_idx", "bin_start")])
tmp_bin_range_filter <- subset(tmp_bin_range, bin_idx %in% c(0, 4, 8, 12, 16, 20, 24, 28, 32))
tmp_bin_range_filter_label <- c("1 kb", "4 kb", "16 kb", "64 kb", "256 kb", "1.024 Mb", "4.096 Mb", "16.38 Mb", "65.54 Mb")
# build the percentage matrix zoom-in view - major
panel_distance_percent_zoom_in_tmp_loose_other_vs_other <- data_distance_percent_smooth_zoom_loose_other_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, major, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, major, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_loose_peak_vs_other <- data_distance_percent_smooth_zoom_loose_peak_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, major, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, major, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_loose_peak_vs_peak <- data_distance_percent_smooth_zoom_loose_peak_vs_peak %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, major, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, major, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_diff_loose_other_vs_other <- panel_distance_percent_zoom_in_tmp_loose_other_vs_other %>%
  #filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, major, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Other vs Other") +
  facet_grid(major ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_loose_peak_vs_other <- panel_distance_percent_zoom_in_tmp_loose_peak_vs_other %>%
  #filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, major, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Other") +
  facet_grid(major ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_loose_peak_vs_peak <- panel_distance_percent_zoom_in_tmp_loose_peak_vs_peak %>%
  #filter(!major %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, major, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Peak") +
  facet_grid(major ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_zoom_in.CTCF.major.pdf", height = 20, width = 10)
plot_grid(panel_distance_percent_zoom_in_diff_loose_other_vs_other, panel_distance_percent_zoom_in_diff_loose_peak_vs_other, panel_distance_percent_zoom_in_diff_loose_peak_vs_peak, nrow = 1)
dev.off()
```


############################
# long / short distribution - ATAC -- manuscript
############################
```{r}
library(ggplot2)
library(ggside)
library(dplyr)
library(tidyr)
library(cowplot)
#######################
# function 
#######################
# load cell annotation
load_cell_annotation <- function(filename) {
  anno_cell <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
  # add major cell type,
  anno_cell <- anno_cell %>%
    mutate(sub2 = sub) %>%
    separate(sub2, sep = ' ', into = c("major", "Extra"), extra = "merge") %>% 
    select(-Extra)
  return(anno_cell)
}
# load sample annotation
load_sample_annotation <- function(filename) {
  anno_sample <- read.table(file = filename, header = T, sep = '\t', stringsAsFactors = F)
}
# load distance track 
load_read_pair_distance_summary_ATAC_xiong <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".pair_distance.ATAC.xiong.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)
    # fix the typo in colnames
    names(table)[names(table) == "sampole_id"] <- "sample_id"
    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
load_read_pair_distance_summary_ATAC_ruochi <- function(folder, sample_list, anno_cell, distance_bin_gruop) {
  # only cell_id included in cell_anno are kept
  data_read_pair_distance_merge = data.frame()
  for (sample_id in sample_list) {
    # get the pair distance file
    pair_distance_summary_file = paste0(folder, "/summary_read_pair_", sample_id, ".pair_distance.ATAC.ruochi.txt")
    # check if file exists
    if (! file.exists(pair_distance_summary_file)) {
      stop(paste0("read pair distance file ", pair_distance_summary_file, " does not exist"))
      return(None)
    }
    # load the file
    table <- read.table(file = pair_distance_summary_file, sep = '\t', header = T, stringsAsFactors = F)
    # fix the typo in colnames
    names(table)[names(table) == "sampole_id"] <- "sample_id"
    # filter the cell id 
    filtered_table <- table %>%
      filter(cell_id %in% anno_cell$cell_id) %>%
      # filter by distance metric
      filter(distance_group == distance_bin_gruop)
    # merge sample
    data_read_pair_distance_merge <- rbind(data_read_pair_distance_merge, filtered_table)
    # logging
    cat(paste0("load pair distance for sample ", sample_id, '\n'))
  }
  # return
  return(data_read_pair_distance_merge)
}
#######################
# load data & annotation
#######################
# cell annotation
anno_cell <- load_cell_annotation("results/202501/cell_meta_table.RNA.tsv")
# sample annotation
anno_sample <- load_sample_annotation("annotation/sample_info.v2.tsv")
# single-cell distance summary
patient_list_20 <- NULL 
# due to privary issue, sample_ids are not provided here
data_pair_distance_mattew_xiong <- load_read_pair_distance_summary_ATAC_xiong("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
data_pair_distance_mattew_ruochi <- load_read_pair_distance_summary_ATAC_ruochi("shared_results/result_GAGE-seq/pipeline_stat/202501/stats", patient_list_20, anno_cell, "Mattew")
#######################
# exploration I-A: erosion score, short-mid-long range
#######################
# --- functions -- 
get_distance_smooth_ATAC <- function(data_pair_distance, agg_ratio = 4) {
  # construct cell by bin data.frame
  tmp <- data_pair_distance %>%
    # remove inter-chromosomal interaction
    filter(bin_idx != "inter") %>% 
    mutate(bin_idx = as.numeric(bin_idx)) %>%
    mutate(bin_idx = bin_idx %/% agg_ratio) %>%
    #mutate(bin_group = ifelse(ATAC_group %in% c("Peak-Peak", "Peak-Other"), "Peak", "Other")) %>%
    mutate(bin_group = ATAC_group) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx, bin_group) %>%
    mutate(bin_start = min(bin_start), bin_end = max(bin_end)) %>%
    select(cell_id, sample_id, distance_group, bin_idx, bin_group, bin_start, bin_end, count) %>%
    group_by(cell_id, sample_id, distance_group, bin_idx, bin_group, bin_start, bin_end) %>%
    summarise(count = sum(count)) %>%
    group_by(cell_id, sample_id, distance_group, bin_group) %>%
    mutate(total_count = sum(count)) %>%
    mutate(percent = count / total_count * 100)
  return(tmp)
}
# --- end
# build the distance percentage table - binning by four bins
data_distance_percent_smooth_xiong <- get_distance_smooth_ATAC(data_pair_distance_mattew_xiong, agg_ratio = 4)
data_distance_percent_smooth_ruochi <- get_distance_smooth_ATAC(data_pair_distance_mattew_ruochi, agg_ratio = 4)
#######################
# exploration I-A: erosion score, short-mid-long range
#######################
# --- functions -- 
get_distance_matrix_smooth_erosion_ATAC <- function(data_distance_percent_smooth, sr_start, sr_end, lr_start, lr_end, agg_ratio = 4) {
  sr_start = sr_start * 4 / agg_ratio
  sr_end = sr_end * 4 / agg_ratio 
  lr_start = lr_start * 4 / agg_ratio
  lr_end = lr_end * 4 / agg_ratio 
  mid_start = sr_end + 1
  mid_end = lr_start - 1
  # get the bin_idx to bin range 
  bin_range <- unique(data_distance_percent_smooth[, c("bin_idx", "bin_start", "bin_end")])
  erosion_score_label_ls <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  erosion_score_label_lm <- paste0(round(bin_range[lr_start, "bin_start"]/1000000, 3), "Mb", "-", round(bin_range[lr_end, "bin_end"]/1000000, 3), "Mb", " vs ", '(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')')
  erosion_score_label_ms <- paste0('(', round(bin_range[mid_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[mid_end, "bin_end"]/1000000, 3), "Mb", ')', " vs ", round(bin_range[sr_start, "bin_start"]/1000, 3), "kb", "-", round(bin_range[sr_end, "bin_end"]/1000, 3), "kb")
  # get the erosion x cell id
  data_erosion <- data_distance_percent_smooth %>%
    mutate(range_group = ifelse(bin_idx >= sr_start - 1 & bin_idx <= sr_end - 1, 'short_range', 
                                ifelse(bin_idx >= mid_start - 1 & bin_idx <= mid_end - 1, 'mid_range',
                                       ifelse(bin_idx >= lr_start - 1 & bin_idx <= lr_end -1, 'long_range', 'other')))) %>%
    group_by(cell_id, bin_group, range_group, total_count) %>%
    summarise(percent =  sum(percent)) %>%
    pivot_wider(names_from = "range_group", values_from = "percent") %>%
    # add 1 pseudo count
    mutate(erosion_score_ls = log2((long_range + 0.1)/(short_range + 0.1)), 
           erosion_score_lm = log2((long_range + 0.1)/(mid_range + 0.1)),
           erosion_score_ms = log2((mid_range + 0.1)/(short_range + 0.1))) %>%
    group_by(bin_group) %>%
    mutate(scaled_erosion_score_ls = scale(erosion_score_ls),
           scaled_erosion_score_lm = scale(erosion_score_lm),
           scaled_erosion_score_ms = scale(erosion_score_ms))
  # add erosion label
  data_erosion$erosion_label_ls <- erosion_score_label_ls
  data_erosion$erosion_label_lm <- erosion_score_label_lm
  data_erosion$erosion_label_ms <- erosion_score_label_ms
  return(data_erosion)
}
# --- end
data_distance_erosion_smooth_xiong <- get_distance_matrix_smooth_erosion_ATAC(data_distance_percent_smooth_xiong, 1, 14, 25, 32, agg_ratio = 4)
data_distance_erosion_smooth_ruochi <- get_distance_matrix_smooth_erosion_ATAC(data_distance_percent_smooth_ruochi, 1, 14, 25, 32, agg_ratio = 4)
# merge with cell annotation
data_distance_erosion_smooth_cell_xiong <- merge(data_distance_erosion_smooth_xiong, anno_cell, by = "cell_id")
data_distance_erosion_smooth_cell_ruochi <- merge(data_distance_erosion_smooth_ruochi, anno_cell, by = "cell_id")
# save the results to files
write.table(data_distance_erosion_smooth_cell_xiong, file = "results/202501/cool_file/distance_erosion_smooth.ATAC.xiong.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
write.table(data_distance_erosion_smooth_cell_ruochi, file = "results/202501/cool_file/distance_erosion_smooth.ATAC.ruochi.tsv", sep = '\t', col.names = T, row.names = F, quote = F)
data_distance_erosion_smooth_cell_xiong <- read.table(file = "results/202501/cool_file/distance_erosion_smooth.ATAC.xiong.tsv", sep = '\t', header = T, stringsAsFactors = F)
data_distance_erosion_smooth_cell_ruochi <- read.table(file = "results/202501/cool_file/distance_erosion_smooth.ATAC.ruochi.tsv", sep = '\t', header = T, stringsAsFactors = F)
########################
# exploration II-A: distance vs contact frequency cluster boxplot (sub-cell type)
########################
# step 1: get the overall long vs short range regardless of disease_group
global_erosion_score_rank_major <- data_distance_erosion_smooth_cell_xiong %>%
  group_by(major) %>%
  filter(bin_group == "Other-Other") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
global_erosion_score_rank_sub <- data_distance_erosion_smooth_cell_xiong %>%
  group_by(sub) %>%
  filter(bin_group == "Other-Other") %>%
  summarise(erosion_score_ave = mean(erosion_score_ls)) %>%
  arrange(erosion_score_ave)
# step 2: for AD and CT rank the cells by global rank major then by global rank sub
global_raw_erosion_ad <- data_distance_erosion_smooth_cell_xiong %>%
  filter(disease_group == 'AD') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(bin_group, major, sub, erosion_score_ls)
global_raw_erosion_ct <- data_distance_erosion_smooth_cell_xiong %>%
  filter(disease_group == 'CT') %>%
  mutate(major = factor(major, levels = global_erosion_score_rank_major$major)) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  arrange(bin_group, major, sub, erosion_score_ls)
# group by disease_group, then by sub-cell type
panel_global_erosion_long_range_boxplot_cell_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = long_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("ATAC-seq peak called by Ruochi") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_mid_range_boxplot_cell_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = mid_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of middle-range interaction") +
  ggtitle("ATAC-seq peak called by Ruochi") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_short_range_boxplot_cell_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = short_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("ATAC-seq peak called by Ruochi") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_long_range_boxplot_cell_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = long_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_mid_range_boxplot_cell_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = mid_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of middle-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_short_range_boxplot_cell_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  group_by(disease_group, sub) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  ggplot(aes(x = bin_group, y = short_range/100, fill = sub, alpha = disease_group)) +
  geom_violin(scale = "width") +
  geom_boxplot(outlier.shape = NA, width = 0.4, position = position_dodge(width = 0.9), fill = "white") +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  scale_fill_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                          "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                          "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                          "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                          "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                          "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                          "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  facet_wrap(~ sub, ncol = 5, scales = "free") +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(strip.background = element_blank()) +
  theme(strip.clip = 'off') +
  theme(legend.position = "none")
panel_global_erosion_long_range_boxplot_patient_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(long_range_ave = mean(long_range)) %>%
  ggplot(aes(x = disease_group, y = long_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_middle_range_boxplot_patient_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(mid_range_ave = mean(mid_range)) %>%
  ggplot(aes(x = disease_group, y = mid_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of mid-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_short_range_boxplot_patient_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(short_range_ave = mean(short_range)) %>%
  ggplot(aes(x = disease_group, y = short_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("ATAC-seq peak in Xiong et al") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_long_range_boxplot_patient_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(long_range_ave = mean(long_range)) %>%
  ggplot(aes(x = disease_group, y = long_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of long-range interaction") +
  ggtitle("ATAC-seq peak called by ruochi") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_middle_range_boxplot_patient_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(mid_range_ave = mean(mid_range)) %>%
  ggplot(aes(x = disease_group, y = mid_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of mid-range interaction") +
  ggtitle("ATAC-seq peak called by ruochi") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
panel_global_erosion_short_range_boxplot_patient_ruochi <- data_distance_erosion_smooth_cell_ruochi %>%
  filter(!sub %in% c("Exc L6 IT", "Exc L5 ET")) %>%
  mutate(sub = factor(sub, levels = global_erosion_score_rank_sub$sub)) %>%
  group_by(bin_group, disease_group, projid, gender, sub) %>%
  summarise(short_range_ave = mean(short_range)) %>%
  ggplot(aes(x = disease_group, y = short_range_ave/100, color = sub, shape = disease_group)) +
  geom_boxplot(outlier.shape = NA, fill = "white", color = "grey") +
  geom_point(alpha = 0.8) +
  xlab("") +
  ylab("Percent of short-range interaction") +
  ggtitle("ATAC-seq peak called by ruochi") +
  facet_grid(bin_group ~ sub, scales = "free") +
  scale_color_manual(name = "", values = c("Astro" = "#665d48", "Endo" = "#8e6c62", "Exc L2/3 IT" = "#b4d335", "Exc L4 IT" = "#61c4b2", 
                                           "Exc L5 ET" = "#0c5b79", "Exc L5 IT" = "#4eb2ad", "Exc L5/6 NP" = "#3c9e64", "Exc L6 CT" = "#2d8db9", 
                                           "Exc L6 IT" = "#a19935", "Exc L6 IT Car3" = "#5352a3", "Exc L6b" = "#6f4a9e", 
                                           "Inh Chandelier" = "#e94b9b", "Inh Lamp5" = "#da808d", "Inh PAX6" = "#702b8b", "Inh Pvalb" = "#d93037", 
                                           "Inh Sncg" = "#b57bb5", "Inh Sst" = "#f8981d", "Inh Vip" = "#9a62a8",  
                                           "Micro" = "#95ae96", "Oligo" = "#53776c", "OPC" = "#384a46", "VLMC" = "#6b7257",
                                           "Exc-Lowexpr" = "#b3cde3", "Oligo-Lowexpr" = "#fbb4ae")) +
  scale_shape_manual(values=c("AD" = 16, "CT" = 4)) +
  scale_y_continuous(labels = scales::percent) +
  theme_cowplot() +
  theme(strip.clip = 'off') +
  theme(strip.text.x = element_text(angle = 45, hjust = 0, vjust = 0)) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(panel.border = element_rect(color = "grey", fill = NA, linewidth = 0.6, linetype = 2)) +
  theme(strip.background = element_blank())
pdf(file = "R_notebook_figures/2025_Feb/figure_overview_erosion_boxplot.ATAC.pdf", height = 12, width = 18)
print(panel_global_erosion_long_range_boxplot_cell_xiong)
print(panel_global_erosion_mid_range_boxplot_cell_xiong)
print(panel_global_erosion_short_range_boxplot_cell_xiong)
print(panel_global_erosion_long_range_boxplot_cell_ruochi)
print(panel_global_erosion_mid_range_boxplot_cell_ruochi)
print(panel_global_erosion_short_range_boxplot_cell_ruochi)
print(panel_global_erosion_long_range_boxplot_patient_xiong)
print(panel_global_erosion_middle_range_boxplot_patient_xiong)
print(panel_global_erosion_short_range_boxplot_patient_xiong)
print(panel_global_erosion_long_range_boxplot_patient_ruochi)
print(panel_global_erosion_middle_range_boxplot_patient_ruochi)
print(panel_global_erosion_short_range_boxplot_patient_ruochi)
dev.off()

########################
# exploration II-D: distance vs contact frequency cluster (smoothed)
########################
# --- functions
get_sorted_distance_distr.cluster <- function(data_distance_percent_smooth, data_distance_erosion_cell, data_distance_erosion_smooth_cluster, bin_group_label) {
  cluster_order_by_erosion <- subset(data_distance_erosion_smooth_cluster, bin_group == bin_group_label)$cluster_id
  # construct cell by bin data.frame
  tmp <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  tmp$cluster_id <- data_distance_erosion_cell$cluster_id[match(tmp$cell_id, data_distance_erosion_cell$cell_id)]
  # build the percentage matrix
  tmp_matrix <- tmp %>% 
    ungroup() %>%
    filter(!is.na(cluster_id)) %>%
    select(cluster_id, bin_idx, count, total_count) %>%
    group_by(cluster_id, bin_idx) %>%
    summarise(count = sum(count), total_count = sum(total_count)) %>%
    mutate(percent = count / total_count * 100) %>%
    select(cluster_id, bin_idx, percent) %>%
    pivot_wider(names_from = "bin_idx", values_from = "percent")
  # sort the rows by cluster order
  tmp_matrix_sorted <- tmp_matrix %>%
    mutate(cluster_id = factor(cluster_id, levels = cluster_order_by_erosion)) %>%
    arrange(cluster_id)
  # confirm the order of cluster id is the same as given
  cat(paste0("The order of cell-cluster match with given order: ", all(tmp_matrix_sorted$cluster_id == cluster_order_by_erosion), "\n"))
  # get percentage matrix
  tmp_mat <- as.matrix(tmp_matrix_sorted[, 2:ncol(tmp_matrix_sorted)])
  #tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  row.names(tmp_mat) <- tmp_matrix_sorted$cluster_id
  # reverse the order of columns
  tmp_mat <- tmp_mat[, ncol(tmp_mat):1]
  #tmp_mat <- t(tmp_mat)
  return(tmp_mat)
}
# --- end
# filter cell types then sort the cell into decile by scaled_erosion_score 
data_distance_erosion_smooth_cell_ntile_xiong <- data_distance_erosion_smooth_cell_xiong %>%
  filter(major %in% c("Astro", "Endo", "Exc", "Inh", "Micro", "Oligo", "OPC", "VLMC")) %>%
  #group_by(erosion_label, sub) %>%
  mutate(erosion_label = erosion_label_ls, erosion_score = erosion_score_ls) %>%
  filter(!is.na(erosion_score)) %>%
  select(bin_group, cell_id, erosion_label, erosion_score, gender, major, sub, disease_group) %>%
  group_by(bin_group, erosion_label, sub, disease_group) %>%
  mutate(erosion_group =  ntile(erosion_score, 20)) %>%
  group_by(bin_group, erosion_label, sub, disease_group, erosion_group) %>%
  mutate(total_count = n()) %>%
  mutate(cluster_id = paste0(sub, '.', disease_group, '.C_', erosion_group))
# cluster
data_distance_erosion_smooth_cluster_xiong <- data_distance_erosion_smooth_cell_ntile_xiong %>%
  group_by(bin_group, erosion_label, sub, disease_group, erosion_group, cluster_id, total_count) %>%
  summarise(cluster_erosion_score = mean(erosion_score)) %>%
  group_by(bin_group, erosion_label, sub, disease_group) %>%
  arrange(cluster_erosion_score, .by_group = T) 
# build the distance percentage table
data_distance_percent_xiong_other_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, data_distance_erosion_smooth_cluster_xiong, 'Other-Other')
data_distance_percent_xiong_peak_vs_other <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, data_distance_erosion_smooth_cluster_xiong, 'Peak-Other')
data_distance_percent_xiong_peak_vs_peak <- get_sorted_distance_distr.cluster(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, data_distance_erosion_smooth_cluster_xiong, 'Peak-Peak')
########################
# plot zoom-in view and aggregated plot - smoothed
########################
# --- functions
build_zoom_in_view <- function(data_distance_percent_smooth, data_distance_erosion_smooth_cell_ntile, bin_group_label) {
  data_distance_percent_smooth <- subset(data_distance_percent_smooth, bin_group == bin_group_label)
  data_distance_erosion_smooth_cell_ntile <- subset(data_distance_erosion_smooth_cell_ntile, bin_group == bin_group_label)
  #
  data_distance_percent_smooth_zoom <- data_distance_percent_smooth
  data_distance_percent_smooth_zoom$cluster_id <- data_distance_erosion_smooth_cell_ntile$cluster_id[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$erosion_group <- data_distance_erosion_smooth_cell_ntile$erosion_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$sub <- data_distance_erosion_smooth_cell_ntile$sub[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$major <- data_distance_erosion_smooth_cell_ntile$major[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$disease_group <- data_distance_erosion_smooth_cell_ntile$disease_group[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  data_distance_percent_smooth_zoom$gender <- data_distance_erosion_smooth_cell_ntile$gender[match(data_distance_percent_smooth_zoom$cell_id, data_distance_erosion_smooth_cell_ntile$cell_id)]
  return(data_distance_percent_smooth_zoom)
}
# --- end
data_distance_percent_smooth_zoom_xiong_other_vs_other <- build_zoom_in_view(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, 'Other-Other')
data_distance_percent_smooth_zoom_xiong_peak_vs_other <- build_zoom_in_view(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, 'Peak-Other')
data_distance_percent_smooth_zoom_xiong_peak_vs_peak <- build_zoom_in_view(data_distance_percent_smooth_xiong, data_distance_erosion_smooth_cell_ntile_xiong, 'Peak-Peak')
#
tmp_bin_range <- unique(data_distance_percent_smooth_zoom_xiong_other_vs_other[, c("bin_idx", "bin_start")])
tmp_bin_range_filter <- subset(tmp_bin_range, bin_idx %in% c(0, 4, 8, 12, 16, 20, 24, 28, 32))
tmp_bin_range_filter_label <- c("1 kb", "4 kb", "16 kb", "64 kb", "256 kb", "1.024 Mb", "4.096 Mb", "16.38 Mb", "65.54 Mb")
# build the percentage matrix zoom-in view - sub
panel_distance_percent_zoom_in_tmp_xiong_other_vs_other <- data_distance_percent_smooth_zoom_xiong_other_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_xiong_peak_vs_other <- data_distance_percent_smooth_zoom_xiong_peak_vs_other %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_tmp_xiong_peak_vs_peak <- data_distance_percent_smooth_zoom_xiong_peak_vs_peak %>% 
  ungroup() %>%
  filter(!is.na(cluster_id)) %>%
  select(cluster_id, erosion_group, sub, disease_group, bin_idx, count, total_count) %>%
  group_by(cluster_id, erosion_group, sub, disease_group, bin_idx) %>%
  summarise(count = sum(count), total_count = sum(total_count)) %>%
  mutate(percent = count / total_count * 100) 
panel_distance_percent_zoom_in_diff_xiong_other_vs_other <- panel_distance_percent_zoom_in_tmp_xiong_other_vs_other %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Other vs Other") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_xiong_peak_vs_other <- panel_distance_percent_zoom_in_tmp_xiong_peak_vs_other %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Other") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
panel_distance_percent_zoom_in_diff_xiong_peak_vs_peak <- panel_distance_percent_zoom_in_tmp_xiong_peak_vs_peak %>%
  filter(!sub %in% c("Exc L5 ET", "Exc L6 IT")) %>%
  ungroup() %>%
  select(erosion_group, sub, bin_idx, disease_group, percent) %>%
  pivot_wider(names_from = "disease_group", values_from = "percent") %>%
  mutate(disease_group = "AD-CT", percent_diff = AD - CT) %>%
  ggplot(aes(y = bin_idx, x = erosion_group, fill = percent_diff)) +
  geom_tile() +
  geom_hline(yintercept = 14, linetype = 2, color = "black") +
  geom_hline(yintercept = 24, linetype = 2, color = "black") +
  geom_hline(yintercept = 32, linetype = 2, color = "black") +
  ylab("Distance group") +
  xlab("Erosion score group") +
  ggtitle("Peak vs Peak") +
  facet_grid(sub ~ disease_group) +
  scale_fill_gradientn(colours = c("#008837", "#f7f7f7", "#7b3294"), values = scales::rescale(c(-1, 0, 1)), guide = "colorbar", limits=c(-1, 1), oob = scales::squish) +
  scale_y_continuous(expand = c(0,0), breaks = tmp_bin_range_filter$bin_idx, labels = tmp_bin_range_filter_label) +
  scale_x_continuous(expand = c(0,0)) +
  theme_cowplot() +
  theme(strip.background = element_blank()) +
  theme(strip.text.y = element_text(angle = 0, hjust = 0)) +
  theme(axis.text.y = element_blank()) +
  theme(axis.line = element_blank()) +
  theme(panel.border = element_rect(color = "black", linewidth = 1)) +
  theme(legend.position = "bottom")
pdf(file = "R_notebook_figures/2025_Feb/figure_circos_zoom_in.ATAC.pdf", height = 50, width = 10)
plot_grid(panel_distance_percent_zoom_in_diff_xiong_other_vs_other, panel_distance_percent_zoom_in_diff_xiong_peak_vs_other, panel_distance_percent_zoom_in_diff_xiong_peak_vs_peak, nrow = 1)
dev.off()
```